---
title: "Long Read IPGs"
author: "Laurel Hiatt"
date: "2024-03-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('dplyr')
library('stringr')
library('ggplot2')
library('cowplot')
library("tidyverse")
library("broom")
library(ggridges)
library(DescTools)
library(Biostrings)
library(plotly)
library(ggbreak) 
library(patchwork)
theme_set(theme_cowplot())
options(stringsAsFactors = FALSE)

calculate_ci <- function(df, successes_col, trials_col, conf_level, method) {

  # Calculate the estimated proportion and confidence intervals for each row
  est_props <- df[[successes_col]] / df[[trials_col]]
  cis <- BinomCI(df[[successes_col]], df[[trials_col]], conf.level = conf_level, method = method)

  # Extract the lower and upper confidence interval bounds
  lower_cis <- cis[, "lwr.ci"]
  upper_cis <- cis[, "upr.ci"]

  # Add the confidence intervals to the original data frame with prefixed names
  df[[paste0(successes_col, "_lower_ci")]] <- lower_cis*100
  df[[paste0(successes_col, "_upper_ci")]] <- upper_cis*100

  # Return the updated data frame
  return(df)
}


replace_list_with_max <- function(df, column_name) {
  df[[column_name]] <- lapply(df[[column_name]], function(lst) {
    if (!is.null(lst) && length(lst) > 0) {
      max_value <- max(as.numeric(lst), na.rm = FALSE)
      return(as.character(max_value))
    } else {
      return(NA)
    }
  })
  return(df)
}



```


```{r normalise}
normalise_str <- function(in_dna) {
  # Split the input by comma if it contains one
  if (grepl(",", in_dna)) {
    in_dna <- unlist(strsplit(in_dna, ",", fixed = TRUE))
  } else {
    in_dna <- list(in_dna)  # Convert single string to list
  }
  
  result <- character(length(in_dna))  # Initialize result vector
  
  for (i in seq_along(in_dna)) {
    dna <- in_dna[i]
    
    if (is.null(dna) || length(dna) == 0) {
      result[i] <- ''  # Return empty string for NULL or empty input
    } else {
      # Generate all circular permutations of input sequence
      all_possible <- sapply(0:(nchar(dna)-1), function(j) 
        paste0(substr(dna, j+1, nchar(dna)), substr(dna, 1, j)))
      
      # Sort permutations alphabetically and return the first
      result[i] <- sort(all_possible)[1]
    }
  }
  
  return(paste(result, collapse = ","))  # Combine result elements with commas
}



# toy <- gnomADSTRcalls[sample(nrow(gnomADSTRcalls), 20), ]
# 
# toy$motif_norm <- sapply(toy$MotifLargest, function(x) normalise_str(as.character(x)))

normalise_call <- function(in_dna) {
  if (is.null(in_dna) || length(in_dna) == 0) {
    return('')
  }
  # Generate all circular permutations of input sequence
  all_possible <- sapply(0:(nchar(in_dna)-1), function(i) paste0(substr(in_dna, i+1, nchar(in_dna)), substr(in_dna, 1, i)))
  # Sort permutations alphabetically and return the first
  return(sort(all_possible)[1])
  }


```

## R Markdown

#Long read
All samples have all loci, so that's nice.
```{r prep longread files}

lr100_motif_counts = read.csv('/Users/quinlan/Downloads/all_samples.motif_counts.tsv', sep = '\t', stringsAsFactors = FALSE)

lr100_motif_counts <- lr100_motif_counts %>%
  rename(gene= Locus)

lr100_motif_counts$gene <- gsub("^.+_", "", lr100_motif_counts$gene)

# Count the number of unique Id values for each unique sample
# sample_id_counts <- lr100_motif_counts %>%
#   group_by(Sample) %>%
#   summarise(Unique_gene_Count = n_distinct(gene))
# 

lr100_motif_edited <- lr100_motif_counts %>%
  select(-c("LocusMotifgene", "Motif"))  %>%
  mutate(AlleleIndex = ifelse(AlleleIndex == 0, "Allele1", 
                               ifelse(AlleleIndex == 1, "Allele2", AlleleIndex)))  

lr100_motif_edited$MotifCount[lr100_motif_edited$MotifCount == "."] <- NA

```

```{r moving around}

lr100_motif_edited <- lr100_motif_edited %>%
  pivot_wider(
    names_from = AlleleIndex,
    values_from = MotifCount, 
    values_fill = list(Allele2LowerBound = NA, Allele1LowerBound = NA)) %>%
    #select(-c(AlleleIndex, Motif, MotifCount)) %>%
    distinct() %>%# Remove unnecessary columns 
    rename(Allele1LowerBound = Allele2) %>%
    rename(Allele2LowerBound = Allele1)




```


#STR table
```{r str}
STR_table <- read.csv('/Users/quinlan/Documents/Git/STRchive/data/STR-disease-loci.csv', stringsAsFactors = FALSE)

#str(STR_table)

#str(gnomADSTRcalls)

STR_table$pathogenic_max[STR_table$pathogenic_max == "11,000"] <- 11000
STR_table$pathogenic_max <- as.integer(STR_table$pathogenic_max)

STR_table$gene[STR_table$gene == "C9orf72"] <- "C9ORF72"

STR_table$gene[STR_table$gene == "ARX" & STR_table$stop_hg38 == 25013697] <- "ARX_1"
STR_table$gene[STR_table$gene == "ARX" & STR_table$stop_hg38 == 25013565] <- "ARX_2"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199966] <- "HOXA13_1"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199861] <- "HOXA13_2"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199732] <- "HOXA13_3"

STR_table_clean <-subset(STR_table, select=c("disease_id", "gene",
                                             "reference_motif_reference_orientation", "Inheritance",
                                             "type", "normal_min", "normal_max",
                                             "intermediate_min", "intermediate_max",
                                             "pathogenic_min", "pathogenic_max",
                                            "pathogenic_motif_reference_orientation"))

STR_table_clean$repeatunit_path_normalized <- sapply(STR_table_clean$pathogenic_motif_reference_orientation, function(x) normalise_str(as.character(x)))


STR_table_clean$repeatunit_ref_normalized <- sapply(STR_table_clean$reference_motif_reference_orientation, function(x) normalise_str(as.character(x)))

#STR_table$repeatunit_ref[STR_table$repeatunit_ref == "GGC/CGC"] <- "NGC"


total <- merge(STR_table_clean,lr100_motif_edited,by="gene")

total <- replace_list_with_max(total, "Allele2LowerBound")
total <- replace_list_with_max(total, "Allele1LowerBound")


total$Allele2LowerBound <- unlist(total$Allele2LowerBound)

total$Allele1LowerBound[sapply(total$Allele1LowerBound, is.null)] <- NA

total$Allele1LowerBound <- unlist(total$Allele1LowerBound)

total <- total %>%
  mutate(
    Allele1LowerBound = as.numeric(Allele1LowerBound),
    Allele2LowerBound = as.numeric(Allele2LowerBound),
    pathogenic_min = as.numeric(pathogenic_min)
  )

```




### For everything, but we're not doing that right now
```{r PG script}

total_prev <- total

#while both inheritances are possible... they're more AD, so, we use that here
total_prev <- total_prev %>%
  mutate(Inheritance = ifelse(gene == 'ATXN2', 'AD', Inheritance))
total_prev <- total_prev %>%
  mutate(Inheritance = ifelse(gene == 'PABPN1', 'AD', Inheritance))

total_prev$repeatunit_path_normalized_list <- strsplit(as.character(total_prev$repeatunit_path_normalized), ",")

AD_total <- subset(total_prev, Inheritance == "AD")
AD_Allele2_greater_than_path_min <- (AD_total$Allele2LowerBound >= AD_total$pathogenic_min) 
AD_result1 <- aggregate(cbind(AD_Allele2_greater_than_path_min), by = list(AD_total$gene), FUN = sum, na.rm=TRUE)
AD_result2 <- aggregate(cbind(AD_total$Allele2LowerBound), by = list(AD_total$gene, AD_total$Inheritance), FUN = function(x) sum(!is.na(x)))
AD_result <- merge(AD_result1, AD_result2)


AR_total <- subset(total_prev, Inheritance == "AR")
# path result if we're NOT doing contraction for VWA1
AR_total <- AR_total %>%
  mutate(pathogenic_min = ifelse(gene == "VWA1", 3, pathogenic_min))
AR_Allele21_greater_than_path_min <- (AR_total$Allele2LowerBound >= AR_total$pathogenic_min) & 
  (AR_total$Allele1LowerBound >= AR_total$pathogenic_min) 
AR_Allele2_greater_than_path_min <- (AR_total$Allele2LowerBound >= AR_total$pathogenic_min) & 
  (AR_total$Allele1LowerBound < AR_total$pathogenic_min) 
AR_result1 <- aggregate(cbind(AR_Allele21_greater_than_path_min), by = list(AR_total$gene), FUN = sum, na.rm=TRUE)
### pathogenic result for VWA1, which is deviation, not contraction
#AR_result1$AR_Allele21_greater_than_path_min[AR_result1$Group.1 == "VWA1"] <- sum(subset(AR_total, gene == "VWA1" & Allele1 != 2 & Allele2 != 2 & (AR_total$motif_norm == AR_total$repeatunit_path_normalized))$gene == "VWA1")
AR_result2 <- aggregate(cbind(AR_Allele2_greater_than_path_min), by = list(AR_total$gene), FUN = sum, na.rm=TRUE)
### carrier result for VWA1, which is deviation, not contraction
#AR_result2$AR_Allele2_greater_than_path_min[AR_result2$Group.1 == "VWA1"] <- sum(subset(AR_total, gene == "VWA1" & (Allele1 == 2 & Allele2 != 2) | (Allele1 != 2 & Allele2 == 2) & (AR_total$motif_norm == AR_total$repeatunit_path_normalized))$gene == "VWA1")
AR_result3 <- aggregate(cbind(AR_total$Allele2), by = list(AR_total$gene, AR_total$Inheritance), FUN = function(x) sum(!is.na(x)))
AR_result <- merge(AR_result1, AR_result2)
AR_result <- merge(AR_result, AR_result3)

XR_XX_total_sex <- subset(total_prev, Inheritance == "XR" & Sex == "XX")
XR_XX_Allele21_greater_than_path_min <- (XR_XX_total_sex$Allele1LowerBound >= XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$Allele2LowerBound >= XR_XX_total_sex$pathogenic_min) 
XR_XX_Allele2_greater_than_path_min <- (XR_XX_total_sex$Allele1LowerBound < XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$Allele2LowerBound >= XR_XX_total_sex$pathogenic_min) 
XR_XX_result1 <- aggregate(cbind(XR_XX_Allele21_greater_than_path_min), by = list(XR_XX_total_sex$gene), FUN = sum, na.rm=TRUE)
XR_XX_result2 <- aggregate(cbind(XR_XX_Allele2_greater_than_path_min), by = list(XR_XX_total_sex$gene), FUN = sum, na.rm=TRUE)
XR_XX_result3 <- aggregate(cbind(XR_XX_total_sex$Allele1), by = list(XR_XX_total_sex$gene,XR_XX_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XR_XX_result <- merge(XR_XX_result1, XR_XX_result2)
XR_XX_result <- merge(XR_XX_result, XR_XX_result3)

XR_XY_total_sex <- subset(total_prev, Inheritance == "XR" & Sex == "XY")
XR_XY_Allele1_greater_than_path_min <- (XR_XY_total_sex$Allele1LowerBound >= XR_XY_total_sex$pathogenic_min) 
XR_XY_Allele2_greater_than_path_min <- XR_XY_total_sex$Allele2LowerBound >= XR_XY_total_sex$pathogenic_min 
XR_XY_Allele_greater_than_path_min <- XR_XY_Allele1_greater_than_path_min | XR_XY_Allele2_greater_than_path_min
XR_XY_result1 <- aggregate(cbind(XR_XY_Allele_greater_than_path_min), by = list(XR_XY_total_sex$gene), FUN = sum, na.rm=TRUE)
XR_XY_result2 <- aggregate(cbind(XR_XY_total_sex$Allele2LowerBound), by = list(XR_XY_total_sex$gene, XR_XY_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XR_XY_result <- merge(XR_XY_result1, XR_XY_result2)

XD_total_sex <- subset(total_prev, Inheritance == "XD")
XD_Allele_greater_than_path_min <- (XD_total_sex$Allele1LowerBound >= XD_total_sex$pathogenic_min)  | 
                                      ((XD_total_sex$Allele2LowerBound >= XD_total_sex$pathogenic_min)) 
XD_result1 <- aggregate(cbind(XD_Allele_greater_than_path_min), by = list(XD_total_sex$gene), FUN = sum, na.rm=TRUE)
XD_result2 <- aggregate(cbind(XD_total_sex$Allele1LowerBound), by = list(XD_total_sex$gene, XD_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XD_result <- merge(XD_result1, XD_result2)


colnames(AD_result) <- c("gene", "Pathogenic_Count", "Inheritance", "Total_Loci")
colnames(AR_result) <- c("gene", "Pathogenic_Count", "Carrier_Count", "Inheritance", "Total_Loci")
colnames(XR_XX_result) <- c("gene", "Pathogenic_Count", "Carrier_Count", "Inheritance", "Total_Loci")
colnames(XR_XY_result) <- c("gene", "Pathogenic_Count", "Inheritance", "Total_Loci")
colnames(XD_result) <- c("gene", "Pathogenic_Count","Inheritance", "Total_Loci")

XR_XY_result$gene <- paste0("XY_", XR_XY_result$gene)
XR_XX_result$gene <- paste0("XX_", XR_XX_result$gene)

AD_result <- transform(AD_result, Carrier_Count = NA)
XR_XY_result <- transform(XR_XY_result, Carrier_Count = NA)
XD_result <- transform(XD_result, Carrier_Count = NA)

# combine data frames
combined_df <- do.call(rbind, list(AD_result, AR_result, XR_XX_result, XR_XY_result, XD_result))

combined_df$pathogenic_percent <- combined_df$Pathogenic_Count/combined_df$Total_Loci*100

combined_df$carrier_percent <- combined_df$Carrier_Count/combined_df$Total_Loci*100

combined_df <- calculate_ci(combined_df, successes_col = "Pathogenic_Count", trials_col = "Total_Loci", conf_level = 0.95)
combined_df <- calculate_ci(combined_df, successes_col = "Carrier_Count", trials_col = "Total_Loci", conf_level = 0.95)


combined_df <- combined_df %>%
  rename(gene = gene)
```


### For DMD
``` {r DMD }


```
## Including Plots

You can also embed plots, for example:

```{r pressure, fig.height = 10}
ggplot(total, aes(x = factor(gene), y = Allele2LowerBound)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
  geom_violin(fill = "black", scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$Id))),
              alpha = 1) +
  labs(title = "Allelic Distribution by Gene",
       x = "gene",
       y = "Allele2 (repeat units)") +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45))
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
