---
title: "Resources"
layout: "@/layouts/Layout.astro"
---

import { FaCompassDrafting } from "react-icons/fa6";
import { PiBooksBold } from "react-icons/pi";
import Link from "@/components/Link";
import styles from "./_resources.module.css";

<section>

# <PiBooksBold/>Resources

</section>

<section>
  <div class={styles.grid}>

    <div>
      ## Computational Methods

      A selection of computational methods that we have either developed or
      used in our research.
      For a more comprehensive review of methods, see <Link to="https://doi.org/10.1038/s41576-024-00692-3">
      Tanudisastro et al. (2024)</Link>.
    </div>

    <div>
      ### Short-read genotyping

      - <Link to="https://github.com/Illumina/ExpansionHunter">ExpansionHunter</Link>
      - <Link to="https://github.com/gymrek-lab/HipSTR">HipSTR</Link>
      - <Link to="https://github.com/gymreklab/gangstr">gangSTR</Link>
    </div>

    <div>
      ### Short-read novel locus detection

      - <Link to="https://github.com/quinlan-lab/STRling">STRling</Link>
      - <Link to="https://github.com/Illumina/ExpansionHunterDenovo">ExpansionHunter Denovo</Link>
    </div>

    <div>
      ### Long-read genotyping

      - <Link to="https://github.com/PacificBiosciences/trgt">TRGT</Link>
      - <Link to="https://github.com/PacificBiosciences/trgt-denovo">TRGT-denovo</Link>
    </div>

  </div>
</section>

<section>
  <div class={styles.grid}>

    <div>
      ## Disease Repeat Catalogs

      - <Link to="/loci/#downloads">STRchive disease loci: TRGT, extended BED and JSON formats</Link>
      - <Link to="https://github.com/broadinstitute/str-analysis/tree/main/str_analysis/variant_catalogs">gnomAD disease loci: ExpansionHunter and TRGT formats</Link>
      - <Link to="https://stripy.org/expansionhunter-catalog-creator">STRipy disease loci: ExpansionHunter format</Link>
    </div>

    <div>
      ## Genome Repeat Catalogs

      - <Link to="https://zenodo.org/records/13178746">PlatinumTRs: 7.8 million reference TRs in TRGT format</Link>
      - <Link to="https://github.com/Illumina/RepeatCatalogs">174,293 polymorphic STRs: ExpansionHunter format</Link>
      - <Link to="https://github.com/PacificBiosciences/trgt/blob/main/repeats/repeat_catalog.hg38.bed">174,293 polymorphic STRs: TRGT format</Link>
    </div>

    <div>
      ## Databases

      - <Link to="https://gnomad.broadinstitute.org/short-tandem-repeats">gnomAD: Tandem Repeats</Link>
      - <Link to="https://stripy.org/">STRipy</Link>
      - <Link to="https://www.ncbi.nlm.nih.gov/books/NBK535148/">GeneReviews: Tandem Repeat Disorders</Link>
      - <Link to="https://omim.org/">OMIM</Link>
      - <Link to="https://webstr.ucsd.edu/pathogenic">WebSTR pathogenic STRs</Link>
      - <Link to="https://wlcb.oit.uci.edu/TRatlas/">TR-Atlas</Link>
    </div>

  </div>
</section>

<section>

### Why do repeat locus definitions differ between resources?

There are often multiple ways to define a given repeat locus.
For example, defining a locus as a stretch of perfectly repeating motifs tends to result in a narrower locus than strategies allowing for interruptions.
In coding regions, the locus boundary might be chosen to align with the reading frame.

There are often multiple "correct" ways to define a given repeat locus, however careful consideration must be made to the downstream uses of the data.
In particular, genotyping accuracy can be affected by the choice of locus definition. Locus definitions can affect the expected allele size, which in turn may have an impact on how allelic thresholds are defined and determined.

The loci in STRchive were defined to be broader, allowing for some interruptions, and with consideration to the biological context and clinical utility of the locus.
This strategy increases the chance that the locus will overlap with a relevant variant call when STRchive is used to annotate a VCF file.
It is also the preferred approach to defining loci for improved genotyping accuracy with <Link to="https://github.com/PacificBiosciences/trgt">TRGT</Link>.
In contrast, <Link to="https://github.com/Illumina/ExpansionHunter">ExpansionHunter</Link> tends to perform better with narrower loci that exclude repeat interruptions.
For this reason, repeat definitions used in <Link to="https://gnomad.broadinstitute.org/short-tandem-repeats">gnomAD</Link> tend to be narrower than those used in STRchive.

#### Defining loci in STRchive

We followed a consistent protocol to arrive at the coordinates within our repeat locus definitions.
We began with the region defined in the literature and used TRF annotations or similar to identify or refine the coordinates if needed.
We examined the reference genome at these coordinates to find the longest exact stretch of the motif (allowing for "N" bases if relevant, e.g. polyalanine loci) and included partial motifs, i.e., the total length does not need to be a multiple of the motif length.
We included interrupted motifs if described in the literature or where they occurred between stretches of the same motif. We defined the coordinates in the hg38 reference genome first and then lifted over to the other genomes.
We conducted a manual review to ensure the coordinates in all genome builds met the above criteria, adjusting as necessary.
To generate TRGT repeat definitions, we further extended the above coordinates to include flanking motifs.


</section>

<section>

### Blueprint for STR evaluation/interpretation

<p class="center">With current resources relevant to each point.</p>

<Link to="/blueprint.html" newTab className="button">
  <FaCompassDrafting />
  Blueprint
</Link>

</section>
