---
title: "gnomADSTRanalysis"
author: "Laurel Hiatt"
date: "2023-04-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('dplyr')
library('stringr')
library('ggplot2')
library('cowplot')
library("tidyverse")
library("broom")
library(ggridges)
library(DescTools)
library(Biostrings)
library(plotly)
library(ggbreak) 
library(patchwork)
theme_set(theme_cowplot())
options(stringsAsFactors = FALSE)

calculate_ci <- function(df, successes_col, trials_col, conf_level, method) {

  # Calculate the estimated proportion and confidence intervals for each row
  est_props <- df[[successes_col]] / df[[trials_col]]
  cis <- BinomCI(df[[successes_col]], df[[trials_col]], conf.level = conf_level, method = method)

  # Extract the lower and upper confidence interval bounds
  lower_cis <- cis[, "lwr.ci"]
  upper_cis <- cis[, "upr.ci"]

  # Add the confidence intervals to the original data frame with prefixed names
  df[[paste0(successes_col, "_lower_ci")]] <- lower_cis*100
  df[[paste0(successes_col, "_upper_ci")]] <- upper_cis*100

  # Return the updated data frame
  return(df)
}

```

## This is where I will put cool explanations of things

### Summary of STR table

### Summary of gnomADSTRcalls

https://gnomad.broadinstitute.org/news/2022-01-the-addition-of-short-tandem-repeat-calls-to-gnomad/
```{r import}
### full data set that is public
#gnomADSTRcalls = read.csv('/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__2022_01_20.tsv', sep = '\t', stringsAsFactors = FALSE)
### data subset to just 1000G, sent by Ben
# gnomADSTRcalls = read.csv('/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__2023_06_15.tsv', sep = '\t', stringsAsFactors = FALSE)

gnomADSTRcalls = read.csv('/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__including_all_age_and_pcr_info__2023_06_28.tsv', sep = '\t', stringsAsFactors = FALSE)

#STR_table <- read.csv('/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/STRpapercode/20240305_STR-disease-loci.processed.csv', stringsAsFactors = FALSE)

#STR_table <- read.csv('/Users/quinlan/Documents/Git/STRchive/data/STR-disease-loci.processed.csv', stringsAsFactors = FALSE)

STR_table <- read.csv('/Users/quinlan/Documents/Git/STRchive/data/STR-disease-loci.csv', stringsAsFactors = FALSE)

#str(STR_table)

#str(gnomADSTRcalls)

STR_table$pathogenic_max[STR_table$pathogenic_max == "11,000"] <- 11000
STR_table$pathogenic_max <- as.integer(STR_table$pathogenic_max)
```


## This clean up helps the STR resource table match the gnomad data
``` {r cleanup}
STR_table$gene[STR_table$gene == "C9orf72"] <- "C9ORF72"

STR_table$gene[STR_table$gene == "ARX" & STR_table$stop_hg38 == 25013697] <- "ARX_1"
STR_table$gene[STR_table$gene == "ARX" & STR_table$stop_hg38 == 25013565] <- "ARX_2"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199966] <- "HOXA13_1"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199861] <- "HOXA13_2"
STR_table$gene[STR_table$gene == "HOXA13" & STR_table$stop_hg38 == 27199732] <- "HOXA13_3"


#STR_table$repeatunit_ref[STR_table$repeatunit_ref == "GGC/CGC"] <- "NGC"


```

## This is just to check and see what doesn't match
``` {r checkdata}
unique(gnomADSTRcalls$Id) -> genelist
unique(STR_table$gene) -> tablegenelist
sort(tablegenelist) -> tablegenelist


# what are the differences between the table and the gnomad gene list
tablegenelist[!(tablegenelist %in% genelist)]

genelist[!(genelist %in% tablegenelist)]

# what are the same genesbetween the table and the gnomad gene list
tablegenelist[(tablegenelist %in% genelist)]
length(tablegenelist[(tablegenelist %in% genelist)])
# CCG in HTT is non-pathogenic but also sometimes surveyed; there might be some association with disease as founder effect?

```

## Let's figure out the motif stuff
``` {r motif}
# MotifStuff
gnomADSTRcalls[c("Allele1Motif", "Allele2Motif")] <- as.data.frame(do.call(rbind, strsplit(gnomADSTRcalls$Motif, "/")), stringsAsFactors = FALSE)
#####Make MotifLargest

gnomADSTRcalls$MotifLargest <- ifelse(is.na(gnomADSTRcalls$Allele2) | gnomADSTRcalls$Allele1 > gnomADSTRcalls$Allele2, 
                                      gnomADSTRcalls$Allele1Motif, 
                                      gnomADSTRcalls$Allele2Motif)


# Only the AATAG is in the BEAN1 loci so... let's do that
#subset(gnomADSTRcalls, gnomADSTRcalls$Id == "BEAN1" & gnomADSTRcalls$motif_norm == "AATGG")
#subset(gnomADSTRcalls, gnomADSTRcalls$Id == "BEAN1" & gnomADSTRcalls$motif_norm == "AATAG")


normalise_str <- function(in_dna) {
  # Split the input by comma if it contains one
  if (grepl(",", in_dna)) {
    in_dna <- unlist(strsplit(in_dna, ",", fixed = TRUE))
  } else {
    in_dna <- list(in_dna)  # Convert single string to list
  }
  
  result <- character(length(in_dna))  # Initialize result vector
  
  for (i in seq_along(in_dna)) {
    dna <- in_dna[i]
    
    if (is.null(dna) || length(dna) == 0) {
      result[i] <- ''  # Return empty string for NULL or empty input
    } else {
      # Generate all circular permutations of input sequence
      all_possible <- sapply(0:(nchar(dna)-1), function(j) 
        paste0(substr(dna, j+1, nchar(dna)), substr(dna, 1, j)))
      
      # Sort permutations alphabetically and return the first
      result[i] <- sort(all_possible)[1]
    }
  }
  
  return(paste(result, collapse = ","))  # Combine result elements with commas
}



# toy <- gnomADSTRcalls[sample(nrow(gnomADSTRcalls), 20), ]
# 
# toy$motif_norm <- sapply(toy$MotifLargest, function(x) normalise_str(as.character(x)))

normalise_call <- function(in_dna) {
  if (is.null(in_dna) || length(in_dna) == 0) {
    return('')
  }
  # Generate all circular permutations of input sequence
  all_possible <- sapply(0:(nchar(in_dna)-1), function(i) paste0(substr(in_dna, i+1, nchar(in_dna)), substr(in_dna, 1, i)))
  # Sort permutations alphabetically and return the first
  return(sort(all_possible)[1])
  }



```

## Let's merge some data
### We are also going to make some new age columns
### And then we're separating out the genotype confidence intervals

```{r merge}
STR_table_clean <-subset(STR_table, select=c("disease_id", "gene",
                                             "reference_motif_reference_orientation", "Inheritance",
                                             "type", "normal_min", "normal_max",
                                             "intermediate_min", "intermediate_max",
                                             "pathogenic_min", "pathogenic_max",
                                             "repeatunitlen", "age_onset_min","typ_age_onset_min",
                                             "age_onset_max", "typ_age_onset_max", "novel", "pathogenic_motif_reference_orientation"))

colnames(STR_table_clean)[2] = "Id"

STR_table_clean$repeatunit_path_normalized <- sapply(STR_table_clean$pathogenic_motif_reference_orientation, function(x) normalise_str(as.character(x)))


STR_table_clean$repeatunit_ref_normalized <- sapply(STR_table_clean$reference_motif_reference_orientation, function(x) normalise_str(as.character(x)))



gnomADSTRcalls$Age[gnomADSTRcalls$Age == '<20'] <- "03-19"
gnomADSTRcalls$Age[gnomADSTRcalls$Age == '>80'] <- "81-99"

gnomADSTRcalls$AgeMax = as.numeric(str_sub(gnomADSTRcalls$Age,-2,-1)) # get last two char in Age. Assumes all <100

gnomADSTRcalls$AgeMin = as.numeric(str_sub(gnomADSTRcalls$Age,1,2)) # get last two char in Age. Assumes all <100


gnomADSTRcalls[c("Allele1LowerBound", "Allele1HigherBound", "Allele2LowerBound", "Allele2HigherBound")] <- as.data.frame(do.call(rbind, strsplit(gsub("/", "-", gnomADSTRcalls$GenotypeConfidenceInterval), "-")), stringsAsFactors = FALSE)
gnomADSTRcalls[c("Allele1LowerBound", "Allele1HigherBound", "Allele2LowerBound", "Allele2HigherBound")] <- lapply(gnomADSTRcalls[c("Allele1LowerBound", "Allele1HigherBound", "Allele2LowerBound", "Allele2HigherBound")], as.integer)

gnomADSTRcalls[c("Allele1LowerBoundOTR", "Allele1HigherBoundOTR", "Allele2LowerBoundOTR", "Allele2HigherBoundOTR")] <- as.data.frame(do.call(rbind, strsplit(gsub("/", "-", gnomADSTRcalls$GenotypeConfidenceIntervalUsingOfftargetRegions), "-")), stringsAsFactors = FALSE)
gnomADSTRcalls[c("Allele1LowerBoundOTR", "Allele1HigherBoundOTR", "Allele2LowerBoundOTR", "Allele2HigherBoundOTR")] <- lapply(gnomADSTRcalls[c("Allele1LowerBound", "Allele1HigherBoundOTR", "Allele2LowerBoundOTR", "Allele2HigherBoundOTR")], as.integer)


#table(gnomADSTRcalls$Age)

#table(gnomADSTRcalls$AgeMin)
#table(gnomADSTRcalls$AgeMax)
gnomADSTRcalls$AlleleLargest <- pmax(gnomADSTRcalls$Allele1, gnomADSTRcalls$Allele2, na.rm = TRUE)


total <- merge(STR_table_clean,gnomADSTRcalls,by="Id")

total$motif_norm <- sapply(total$MotifLargest, function(x) normalise_call(as.character(x)))
total$motif_norm_small <- sapply(total$Allele1Motif, function(x) normalise_call(as.character(x)))

```


``` {r comparing_norm_motifs}

### regex ? 
table(total$repeatunit_path_normalized == total$motif_norm)
table(total$repeatunit_path_normalized == total$motif_norm | total$repeatunit_ref_normalized == total$motif_norm)


subset(total, repeatunit_path_normalized != motif_norm & repeatunit_ref_normalized != motif_norm,
       select = c("motif_norm", "repeatunit_ref_normalized", "repeatunit_path_normalized", "Id"))

unique(subset(total, repeatunit_path_normalized != motif_norm & repeatunit_ref_normalized != motif_norm,
       select = c("motif_norm", "repeatunit_ref_normalized", "repeatunit_path_normalized", "Id")))

mismatchmotif <- total %>%
  subset(repeatunit_path_normalized != motif_norm & repeatunit_ref_normalized != motif_norm,
         select = c("motif_norm", "repeatunit_ref_normalized", "repeatunit_path_normalized", "Id")) %>%
  count(Id, motif_norm, repeatunit_ref_normalized, repeatunit_path_normalized)

subset(total, total$motif_norm == "ATGTT" & total$Id == "YEATS2")

```


## Let's subset the data based on motif, as discussed with Harriet
Because the CNG motifs seem to be the ones causing the most trouble... let's
not use those.

``` {r no CNG}
total <- subset(total, total$motif_norm != "CNG")

gnomADSTRcalls <- subset(gnomADSTRcalls, gnomADSTRcalls$motif_norm != "CNG")

```


## Loop from harriet for allele distribution with age vs no age available
```{r loop}
for (locus in unique(gnomADSTRcalls$Id)) {
  ggplot(subset(gnomADSTRcalls, Id == locus)) + 
    geom_histogram(aes(x = Allele2, fill = Age), alpha = 0.5) + 
    facet_wrap(~AgeMax == 'age_not_available', scales = 'free_y') +
    ggtitle(locus)
  ggsave(paste0(locus,'_allele2_age_available.png'))
}

for (locus in unique(gnomADSTRcalls$Id)) {
  ggplot(subset(gnomADSTRcalls, Id == locus)) + 
    geom_histogram(aes(x = Allele2UsingOfftargetRegions, fill = Age), alpha = 0.5) + 
    facet_wrap(~AgeMax == '_OTR_age_not_available', scales = 'free_y') +
    ggtitle(locus)
  ggsave(paste0(locus,'_OTR_allele2_age_available.png'))
}

table(gnomADSTRcalls$Age)

wilcox.test(subset(gnomADSTRcalls, Age == 'age_not_available')$Allele2, 
                          subset(gnomADSTRcalls, Age != 'age_not_available')$Allele2)


# results <- data.frame()  # create an empty dataframe to store results
# 
# for (locus in unique(gnomADSTRcalls$Id)) {
#   p <- tidy(ks.test(subset(gnomADSTRcalls, Age == 'age_not_available' & Id == locus)$Allele2, 
#                          subset(gnomADSTRcalls, Age != 'age_not_available'& Id == locus)$Allele2))
#   results <- rbind(results, data.frame(Id = locus, statistic = p$statistic, p.value = p$p.value))
# }
# 
# print(results)

# NOT ENOUGH Y DATA!!!
```

## PCR comparison
``` {r PCR, fig.width=15,fig.height=10}

ggplot(gnomADSTRcalls) + 
  geom_histogram(aes(x = Allele2, fill = PcrProtocol), alpha = 0.5) + 
  facet_wrap(~PcrProtocol, scales = 'free_y') +
  ggtitle('PCR Protocol')

# 
wilcox.test(subset(gnomADSTRcalls, PcrProtocol == 'pcr_free')$Allele2,
             subset(gnomADSTRcalls, PcrProtocol == 'pcr_plus')$Allele2)
p <- tidy(wilcox.test(subset(gnomADSTRcalls, PcrProtocol == 'pcr_free')$Allele2,
             subset(gnomADSTRcalls, PcrProtocol == 'pcr_plus')$Allele2))


# for (locus in unique(gnomADSTRcalls$Id)) {
#   cat(paste0(locus))
#   p <- tidy(wilcox.test(subset(gnomADSTRcalls, PcrProtocol == 'pcr_free' & Id == locus)$Allele2, 
#              subset(gnomADSTRcalls, PcrProtocol == 'pcr_plus'& Id == locus)$Allele2))
#   cat('\n') 
#   cat("The statistic is \n")
#   cat(p$statistic)
#   cat('\n') 
#   cat("The p-value is \n")
#   cat(p$p.value)
#   cat('\n') 
# }

results <- data.frame()  # create an empty dataframe to store results

for (locus in unique(gnomADSTRcalls$Id)) {
  p <- tidy(wilcox.test(subset(gnomADSTRcalls, PcrProtocol == 'pcr_free' & Id == locus)$Allele2, 
                         subset(gnomADSTRcalls, PcrProtocol == 'pcr_plus'& Id == locus)$Allele2))
  results <- rbind(results, data.frame(Id = locus, statistic = p$statistic, p.value = p$p.value))
}

print(results)

ggplot(results, aes(x = Id, y = p.value, color = Id)) +
  geom_point() +
  geom_hline(yintercept = 0.05, linetype = "dashed") +
  labs(x = "Locus", y = "p-value", color = "Locus") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + theme(legend.position = "none")

ggplot(results, aes(x = Id, y = -log10(p.value), color = Id)) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  labs(x = "Locus", y = "-log(p-value)", color = "Locus") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + theme(legend.position = "none") + scale_y_continuous(trans='log10')
```

## Let's look at Allele2 compared to Allele2UsingOfftargetRegions
Consider plotting a loci with Allele2 AND Allele2UsingOffTarget (sanity check)
Scatter (look for outliers in plot, maybe draw line of where we expect error)
Could use Allele2UOT with more confidenceâ€¦

``` {r allele2check, fig.width=15}
ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele2, y = Allele2UsingOfftargetRegions, col=Id)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylim(0,400)
```

``` {r allele1check, fig.width=15}
ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele1, y = Allele1UsingOfftargetRegions, col=Id)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylim(0,400)
```

``` {r allele1vs2check, fig.width=15}
ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele1, y = Allele2, col=Id)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylim(0,400)
```


``` {r allele1vs2check, fig.width=15}
ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele1UsingOfftargetRegions, y = Allele2UsingOfftargetRegions, col=Id)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylim(0,400)
```


``` {r allele2check by loci, fig.width=15,fig.height=100}

ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele2, y = Allele2UsingOfftargetRegions, col=Motif)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = "none") + facet_wrap(~Id, ncol=2, scales = "free") 


ggplot(gnomADSTRcalls) + geom_point(aes(x = AlleleLargest, y = Allele2UsingOfftargetRegions, col=Motif)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = "none") + facet_wrap(Id~motif_norm, ncol=2, scales = "free") + geom_abline(slope=1,intercept=0, col="gray")


ggplot(gnomADSTRcalls) + geom_point(aes(x = AlleleLargest, y = Allele2UsingOfftargetRegions, col=Motif)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = "none") + facet_wrap(~Id, ncol=2, scales = "free") + geom_abline(slope=1,intercept=0, col="gray")
```

``` {r allele1check by loci, fig.width=15,fig.height=100}
ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele1, y = Allele1UsingOfftargetRegions, col=Motif)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = "none") + facet_wrap(~Id, ncol=2, scales = "free") 
```

``` {r allele1vs2check by loci, fig.width=15,fig.height=100}

ggplot(gnomADSTRcalls) + geom_point(aes(x = Allele1, y = Allele2, col=Motif)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = "none") + ylim(0,400) + facet_wrap(~Id, ncol=2, scales = "free") 
```

``` {r lines}
total_Hox1 <- subset(total, Id == 'HOXA13_1') # 3
total_Hox2 <- subset(total, Id == 'HOXA13_2') # 3
total_Hox3 <- subset(total, Id == 'HOXA13_3') # 3
total_phox <- subset(total, Id == 'PHOX2B') # 3
total_tbx1 <- subset(total, Id == 'TBX1') # 3
total_zic2 <- subset(total, Id == 'ZIC2') # 3
total_foxl2 <- subset(total, Id == 'FOXL2') # 3
total_bean1 <- subset(total, Id == 'BEAN1') # 5
total_rapgef2 <- subset(total, Id == 'RAPGEF2') # 5

tail(sort(table(total_Hox1$Allele2)), 3)
tail(sort(table(total_Hox2$Allele2)), 3)
tail(sort(table(total_Hox3$Allele2)), 3)
tail(sort(table(total_phox$Allele2)), 3)
tail(sort(table(total_tbx1$Allele2)), 3)
tail(sort(table(total_zic2$Allele2)), 3)
tail(sort(table(total_foxl2$Allele2)), 3)
tail(sort(table(total_bean1$Allele2)), 3)
tail(sort(table(total_rapgef2$Allele2)), 3)

unique(total_Hox1$End - total_Hox1$Start_0based)/3
unique(total_Hox2$End - total_Hox2$Start_0based)/3
unique(total_Hox3$End - total_Hox3$Start_0based)/3
unique(total_phox$End - total_phox$Start_0based)/3
unique(total_tbx1$End - total_tbx1$Start_0based)/3
unique(total_zic2$End - total_zic2$Start_0based)/3
unique(total_foxl2$End - total_foxl2$Start_0based)/3
unique(total_bean1$End - total_bean1$Start_0based)/5
unique(total_rapgef2$End - total_rapgef2$Start_0based)/5

# conclusion: 3bp repeats are defaulting to range of loci, 5bp are not, they're just messy


#total_dmd <- subset(total, Id == 'DMD') # 5
#table(total_dmd$Sex)
# XX    XY 
# 8167 10970 
# XYs for X-linked inheritance are filtered out!!!!!


#table(total_Hox1$Motif)
#table(total_Hox2$Motif)
#table(total_Hox3$Motif)


```

# Here we have the histograms, and there is a lot of overlap. So it is ugly.
https://stackoverflow.com/questions/6957549/overlaying-histograms-with-ggplot2-in-r
``` {r histograms, fig.width=15,fig.height=100}
ggplot(gnomADSTRcalls) + geom_histogram(aes(x = Allele1, fill = "Allele1", alpha = 0.2)) + geom_histogram(aes(x = Allele2, fill = "Allele2", alpha = 0.2)) + geom_histogram(aes(x = Allele1UsingOfftargetRegions, fill = "Allele1UsingOfftargetRegions", alpha = 0.2)) + geom_histogram(aes(x = Allele2UsingOfftargetRegions, fill = "Allele2UsingOfftargetRegions", alpha = 0.2)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + facet_wrap(~Id, ncol=2, scales = "free") 

ggplot(gnomADSTRcalls) + geom_density(aes(x = Allele1, color = "Allele1", alpha = 0.2)) + geom_density(aes(x = Allele2, color = "Allele2", alpha = 0.2)) + geom_density(aes(x = Allele1UsingOfftargetRegions, color = "Allele1UsingOfftargetRegions", alpha = 0.2)) + geom_density(aes(x = Allele2UsingOfftargetRegions, color = "Allele2UsingOfftargetRegions", alpha = 0.2)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + facet_wrap(~Id, ncol=2, scales = "free") 

```

## Let's see if variants fall into type by category
``` {r type}
ggplot(total) + geom_point(aes(x = Id, y = Allele2UsingOfftargetRegions, col=type)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) + facet_grid(~type, scales = "free_x") + ylim(0,4000) + scale_x_discrete(guide = guide_axis(check.overlap = TRUE))

# ggplot(total) + geom_point(aes(x = Id, y = Allele2UsingOfftargetRegions, col=type)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) + facet_grid(~type+IsNeuroCase, scales = "free_x") + ylim(0,4000) + scale_x_discrete(guide = guide_axis(check.overlap = TRUE))

```

``` {r type, fig.width = 20}
ggplot(total) + geom_point(aes(x = Id, y = Allele2, col=motif_norm=="CNG")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) + facet_grid(~type, scales = "free_x") + ylim(0,400) + scale_x_discrete(guide = guide_axis(check.overlap = TRUE))

```


``` {r type}
ggplot(total) + geom_point(aes(x = Id, y = AlleleLargest, col=type)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) + facet_grid(~type, scales = "free_x") + ylim(0,400) + scale_x_discrete(guide = guide_axis(check.overlap = TRUE))

```

## Question 1
Allele size beyond pathogenic in literature
Penetrance lower than expected?

```{r Q1}
table(total$Allele2 >=  total$pathogenic_max, useNA = "ifany")
table(total$Allele2UsingOfftargetRegion >=  total$pathogenic_max)

# we care most about the min 
table(total$Allele2 >=  total$pathogenic_min)
table(total$Allele2UsingOfftargetRegion >=  total$pathogenic_min)

#Subs1<-subset(total, (!is.na(total$Allele2)) & (!is.na(total$pathogenic_max)))
#table(Subs1$Allele2 >=  Subs1$pathogenic_max)
# it's not NA in these columns that are making the discrepancy
#Allele2_greater_than_path_max <- Subs1$Allele2 > Subs1$pathogenic_max
#result1 <- aggregate(cbind(Allele2_greater_than_path_max), by = list(Subs1$Id), FUN = sum, na.rm=TRUE)

#Subs2<-subset(total, (!is.na(total$Allele2)) & (!is.na(total$pathogenic_max)) & (!is.na(total$Id)))
#table(Subs2$Allele2 >=  Subs2$pathogenic_max)
# it's not the three columns so like... what is it


# Create a logical vector to represent whether x > y for each row
# Allele2_greater_than_path_max <- total$AlleleLargest >= total$pathogenic_max
# table(Allele2_greater_than_path_max)
# 
# # Use aggregate() to count occurrences of x > y by z value
# result1 <- aggregate(cbind(Allele2_greater_than_path_max), by = list(total$Id), FUN = sum, na.rm=TRUE)
# 
# sum(result1$Allele2_greater_than_path_max)
# 
# df_summary <- total %>%
#   group_by(Id) %>%
#   summarize(Allele2_greater_than_path_max = sum(Allele2 >= pathogenic_max, na.rm = TRUE))
# 
# sum(df_summary$Allele2_greater_than_path_max)
# 
# sum(result1$Allele2_greater_than_path_max)
# 
# result2 <- aggregate(cbind(total$Allele2), by = list(total$Id), FUN = function(x) sum(!is.na(x)))
# 
# result <- merge(result1,result2)


#MIN
# Create a logical vector to represent whether x > y for each row
Allele2_greater_than_path_min <- total$Allele2LowerBound >= total$pathogenic_min
table(Allele2_greater_than_path_min)

# Use aggregate() to count occurrences of x > y by z value
result1 <- aggregate(cbind(Allele2_greater_than_path_min), by = list(total$Id), FUN = sum, na.rm=TRUE)

sum(result1$Allele2_greater_than_path_min)

result2 <- aggregate(cbind(total$Allele2), by = list(total$Id, total$repeatunit_path_normalized), FUN = function(x) sum(!is.na(x)))

result <- merge(result1,result2)


# Rename the columns of the result dataframe
colnames(result) <- c("Id", "Allele2>pathogenic_min", "pathogenic_motif", "Total_Allele2Loci_for_Id")

result$percent<- (result$`Allele2>pathogenic_min`/result$Total_Allele2Loci_for_Id)*100



calculate_ci <- function(df, successes_col, trials_col, conf_level, method) {
  
  # Calculate the estimated proportion and confidence intervals for each row
  est_props <- df[[successes_col]] / df[[trials_col]]
  cis <- BinomCI(df[[successes_col]], df[[trials_col]], conf.level = conf_level, method = method)
  
  # Extract the lower and upper confidence interval bounds
  lower_cis <- cis[, "lwr.ci"]
  upper_cis <- cis[, "upr.ci"]
  
  # Add the confidence intervals to the original data frame
  df$lower_ci <- lower_cis*100
  df$upper_ci <- upper_cis*100
  
  # Return the updated data frame
  return(df)
}

result <- calculate_ci(result, successes_col = "Allele2>pathogenic_min", trials_col = "Total_Allele2Loci_for_Id", conf_level = 0.95)

ggplot(result, aes(x = reorder(Id,percent), y = percent, color = pathogenic_motif)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") +
  ggtitle("Pathogenic Percentage by Id with 95% CI")


ggplot(subset(result, percent > 1), aes(x = reorder(Id, percent), y = percent, color = pathogenic_motif)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") + ggtitle("Pathogenic % by Id w/ 95% CI (Pct > 1)")

ggplot(subset(result, percent <= 1), aes(x = reorder(Id, percent), y = percent, color = pathogenic_motif)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") + ggtitle("Pathogenic % by Id w/ 95% CI (Pct<= 1)")

ggplot(subset(result, percent <= 0.1), aes(x = reorder(Id, percent), y = percent, color = pathogenic_motif)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") + ggtitle("Pathogenic % by Id w/ 95% CI (Pct<= 0.1)")


#ggplot(total) +
 # geom_histogram(aes(x = Allele2, fill = Id), alpha = 0.5) + facet_wrap(~Allele2 > pathogenic_max, scales = 'free_y') + theme(legend.position = "none")

#total_pabpn1 <- subset(total, Id == 'PABPN1') # 5
#table(total_pabpn1$Allele2)
```



### Inheritance patterns?
```{r inheritance}

Allele2_greater_than_path_min <- total$Allele2LowerBound >= total$pathogenic_min
table(Allele2_greater_than_path_min)

# Use aggregate() to count occurrences of x > y by z value
result1 <- aggregate(cbind(Allele2_greater_than_path_min) ~ Id + Inheritance, data = total, FUN = sum, na.rm = TRUE)

sum(result1$Allele2_greater_than_path_min)

result2 <- aggregate(cbind(Allele2_greater_than_path_min) ~ Id + Inheritance, data = total, FUN = function(x) sum(!is.na(x)))

merge(result1,result2)

inheritance_result <- result1 %>% inner_join( result2, 
        by=c('Id', "Inheritance"))

# Rename the columns of the result dataframe
colnames(inheritance_result) <- c("Id", "Inheritance", "Allele2>pathogenic_min", "Total_Allele2Loci_for_Id")

inheritance_result$percent<- (inheritance_result$`Allele2>pathogenic_min`/inheritance_result$Total_Allele2Loci_for_Id)*100

sum(inheritance_result$Allele2_greater_than_path_min, na.rm= TRUE)

# https://cran.r-project.org/web/packages/interpretCI/vignettes/Confidence_interval_for_a_proportion.html
# https://rcompanion.org/handbook/H_02.html

# calculate_ci <- function(df, successes_col, trials_col, conf_level, method) {
#   
#   # Calculate the estimated proportion and confidence intervals for each row
#   est_props <- df[[successes_col]] / df[[trials_col]]
#   cis <- BinomCI(df[[successes_col]], df[[trials_col]], conf.level = conf_level, method = method)
#   
#   # Extract the lower and upper confidence interval bounds
#   lower_cis <- cis[, "lwr.ci"]
#   upper_cis <- cis[, "upr.ci"]
#   
#   # Add the confidence intervals to the original data frame
#   df$lower_ci <- lower_cis*100
#   df$upper_ci <- upper_cis*100
#   
#   # Return the updated data frame
#   return(df)
# }

inheritance_result <- calculate_ci(inheritance_result, successes_col = "Allele2>pathogenic_min", trials_col = "Total_Allele2Loci_for_Id", conf_level = 0.95)

ggplot(inheritance_result, aes(x = reorder(Id,percent), y = percent, color = Inheritance)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") +
  ggtitle("Pathogenic Percentage by Id with 95% CI with Inheritance Info")


```

### Let's see about if we parse these out by PCR status
```{r PCR}

total_PCR <- subset(total, total$PcrProtocol == "pcr_free" | total$PcrProtocol == "pcr_plus")

table(total_PCR$PcrProtocol)

Allele2_greater_than_path_min <- total_PCR$Allele2LowerBound >= total_PCR$pathogenic_min
table(Allele2_greater_than_path_min)

# Use aggregate() to count occurrences of x > y by z value
result1 <- aggregate(cbind(Allele2_greater_than_path_min) ~ Id + PcrProtocol, data = total_PCR, FUN = sum, na.rm = TRUE)

result2 <- aggregate(cbind(Allele2_greater_than_path_min) ~ Id + PcrProtocol, data = total_PCR, FUN = function(x) sum(!is.na(x)))

merge(result1,result2)

PCR_result <- result1 %>% inner_join( result2, 
        by=c('Id', "PcrProtocol"))

# Rename the columns of the result dataframe
colnames(PCR_result) <- c("Id", "PcrProtocol", "Allele2>pathogenic_min", "Total_Allele2Loci_for_Id")

PCR_result$percent<- (PCR_result$`Allele2>pathogenic_min`/PCR_result$Total_Allele2Loci_for_Id)*100

sum(PCR_result$Allele2_greater_than_path_max, na.rm= TRUE)


# https://cran.r-project.org/web/packages/interpretCI/vignettes/Confidence_interval_for_a_proportion.html
# https://rcompanion.org/handbook/H_02.html


calculate_ci <- function(df, successes_col, trials_col, conf_level, method) {

  # Calculate the estimated proportion and confidence intervals for each row
  est_props <- df[[successes_col]] / df[[trials_col]]
  cis <- BinomCI(df[[successes_col]], df[[trials_col]], conf.level = conf_level, method = method)

  # Extract the lower and upper confidence interval bounds
  lower_cis <- cis[, "lwr.ci"]
  upper_cis <- cis[, "upr.ci"]

  # Add the confidence intervals to the original data frame
  df$lower_ci <- lower_cis*100
  df$upper_ci <- upper_cis*100

  # Return the updated data frame
  return(df)
}

PCR_result <- calculate_ci(PCR_result, successes_col = "Allele2>pathogenic_min", trials_col = "Total_Allele2Loci_for_Id", conf_level = 0.95)

#### PLOT WITH PERCENTAGE DATA!!!!!

ggplot(PCR_result, aes(x = reorder(Id,percent), y = percent, color = PcrProtocol)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") +
  ggtitle("Pathogenic Percentage by Id with 95% CI with Protocol Info")


t.test(PCR_result$`Allele2>pathogenic_min` ~ PcrProtocol, data = PCR_result)


# Welch Two Sample t-test
# 
# data:  PCR_result$`Allele2>pathogenic_min` by PcrProtocol
# t = 1.7474, df = 60.672, p-value = 0.08563
# alternative hypothesis: true difference in means between group pcr_free and group pcr_plus is not equal to 0
# 95 percent confidence interval:
#  -15.05869 223.53142
# sample estimates:
# mean in group pcr_free mean in group pcr_plus 
#              146.78182               42.54545 

```


### Let's do the inheritance pattern analysis
```{r inheritance}
total_prev <- subset(total, total$motif_norm != "CNG")

#while both inheritances are possible... they're more AD, so, we use that here
total_prev <- total_prev %>%
  mutate(Inheritance = ifelse(Id == 'ATXN2', 'AD', Inheritance))
total_prev <- total_prev %>%
  mutate(Inheritance = ifelse(Id == 'PABPN1', 'AD', Inheritance))

total_prev <- total_prev %>%
  mutate(repeatunit_path_normalized = ifelse(repeatunit_path_normalized == 'CNG', 'CAG,CCG,CGG,CTG', repeatunit_path_normalized))
total_prev$repeatunit_path_normalized_list <- strsplit(as.character(total_prev$repeatunit_path_normalized), ",")

AD_total <- subset(total_prev, Inheritance == "AD")
AD_Allele2_greater_than_path_min <- (AD_total$Allele2LowerBound >= AD_total$pathogenic_min) & 
  (AD_total$motif_norm %in% AD_total$repeatunit_path_normalized_list)
AD_result1 <- aggregate(cbind(AD_Allele2_greater_than_path_min), by = list(AD_total$Id), FUN = sum, na.rm=TRUE)
AD_result2 <- aggregate(cbind(AD_total$Allele2), by = list(AD_total$Id, AD_total$repeatunit_path_normalized, AD_total$Inheritance), FUN = function(x) sum(!is.na(x)))
AD_result <- merge(AD_result1, AD_result2)


AR_total <- subset(total_prev, Inheritance == "AR")
# path result if we're NOT doing contraction for VWA1
AR_total <- AR_total %>%
  mutate(pathogenic_min = ifelse(Id == "VWA1", 3, pathogenic_min))
AR_Allele21_greater_than_path_min <- (AR_total$Allele2LowerBound >= AR_total$pathogenic_min) & 
  (AR_total$motif_norm %in% AR_total$repeatunit_path_normalized_list) & 
  (AR_total$motif_norm_small %in% AR_total$repeatunit_path_normalized_list) &
  (AR_total$Allele1LowerBound >= AR_total$pathogenic_min) 
AR_Allele2_greater_than_path_min <- (AR_total$Allele2LowerBound >= AR_total$pathogenic_min) & 
  (AR_total$Allele1LowerBound < AR_total$pathogenic_min) & 
  (AR_total$motif_norm %in% AR_total$repeatunit_path_normalized) 
AR_result1 <- aggregate(cbind(AR_Allele21_greater_than_path_min), by = list(AR_total$Id), FUN = sum, na.rm=TRUE)
### pathogenic result for VWA1, which is deviation, not contraction
#AR_result1$AR_Allele21_greater_than_path_min[AR_result1$Group.1 == "VWA1"] <- sum(subset(AR_total, Id == "VWA1" & Allele1 != 2 & Allele2 != 2 & (AR_total$motif_norm == AR_total$repeatunit_path_normalized))$Id == "VWA1")
AR_result2 <- aggregate(cbind(AR_Allele2_greater_than_path_min), by = list(AR_total$Id), FUN = sum, na.rm=TRUE)
### carrier result for VWA1, which is deviation, not contraction
#AR_result2$AR_Allele2_greater_than_path_min[AR_result2$Group.1 == "VWA1"] <- sum(subset(AR_total, Id == "VWA1" & (Allele1 == 2 & Allele2 != 2) | (Allele1 != 2 & Allele2 == 2) & (AR_total$motif_norm == AR_total$repeatunit_path_normalized))$Id == "VWA1")
AR_result3 <- aggregate(cbind(AR_total$Allele2), by = list(AR_total$Id, AR_total$repeatunit_path_normalized, AR_total$Inheritance), FUN = function(x) sum(!is.na(x)))
AR_result <- merge(AR_result1, AR_result2)
AR_result <- merge(AR_result, AR_result3)

XR_XX_total_sex <- subset(total_prev, Inheritance == "XR" & Sex == "XX")
XR_XX_Allele21_greater_than_path_min <- (XR_XX_total_sex$Allele1LowerBound >= XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$Allele2LowerBound >= XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$motif_norm %in% XR_XX_total_sex$repeatunit_path_normalized) &
  (XR_XX_total_sex$motif_norm_small %in% XR_XX_total_sex$repeatunit_path_normalized)
XR_XX_Allele2_greater_than_path_min <- (XR_XX_total_sex$Allele1LowerBound < XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$Allele2LowerBound >= XR_XX_total_sex$pathogenic_min) & 
  (XR_XX_total_sex$motif_norm %in% XR_XX_total_sex$repeatunit_path_normalized)
XR_XX_result1 <- aggregate(cbind(XR_XX_Allele21_greater_than_path_min), by = list(XR_XX_total_sex$Id), FUN = sum, na.rm=TRUE)
XR_XX_result2 <- aggregate(cbind(XR_XX_Allele2_greater_than_path_min), by = list(XR_XX_total_sex$Id), FUN = sum, na.rm=TRUE)
XR_XX_result3 <- aggregate(cbind(XR_XX_total_sex$Allele1), by = list(XR_XX_total_sex$Id, XR_XX_total_sex$repeatunit_path_normalized, XR_XX_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XR_XX_result <- merge(XR_XX_result1, XR_XX_result2)
XR_XX_result <- merge(XR_XX_result, XR_XX_result3)

XR_XY_total_sex <- subset(total_prev, Inheritance == "XR" & Sex == "XY")
XR_XY_Allele1_greater_than_path_min <- (XR_XY_total_sex$Allele1LowerBound >= XR_XY_total_sex$pathogenic_min) &
  (XR_XY_total_sex$motif_norm %in% XR_XY_total_sex$repeatunit_path_normalized_list)
XR_XY_Allele2_greater_than_path_min <- XR_XY_total_sex$Allele2LowerBound >= XR_XY_total_sex$pathogenic_min &
  (XR_XY_total_sex$motif_norm %in% XR_XY_total_sex$repeatunit_path_normalized_list)
XR_XY_Allele_greater_than_path_min <- XR_XY_Allele1_greater_than_path_min | XR_XY_Allele2_greater_than_path_min
XR_XY_result1 <- aggregate(cbind(XR_XY_Allele_greater_than_path_min), by = list(XR_XY_total_sex$Id), FUN = sum, na.rm=TRUE)
XR_XY_result2 <- aggregate(cbind(XR_XY_total_sex$Allele1), by = list(XR_XY_total_sex$Id, XR_XY_total_sex$repeatunit_path_normalized, XR_XY_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XR_XY_result <- merge(XR_XY_result1, XR_XY_result2)

XD_total_sex <- subset(total_prev, Inheritance == "XD")
XD_Allele_greater_than_path_min <- ((XD_total_sex$Allele1LowerBound >= XD_total_sex$pathogenic_min) &
                                      (XD_total_sex$motif_norm_small %in% XD_total_sex$repeatunit_path_normalized_list)) | 
                                      ((XD_total_sex$Allele2LowerBound >= XD_total_sex$pathogenic_min)) & 
  (XD_total_sex$motif_norm %in% XD_total_sex$repeatunit_path_normalized_list)
XD_result1 <- aggregate(cbind(XD_Allele_greater_than_path_min), by = list(XD_total_sex$Id), FUN = sum, na.rm=TRUE)
XD_result2 <- aggregate(cbind(XD_total_sex$Allele1), by = list(XD_total_sex$Id, XD_total_sex$repeatunit_path_normalized, XD_total_sex$Inheritance), FUN = function(x) sum(!is.na(x)))
XD_result <- merge(XD_result1, XD_result2)


colnames(AD_result) <- c("Id", "Pathogenic_Count", "Motif", "Inheritance", "Total_Loci")
colnames(AR_result) <- c("Id", "Pathogenic_Count", "Carrier_Count", "Motif", "Inheritance", "Total_Loci")
colnames(XR_XX_result) <- c("Id", "Pathogenic_Count", "Carrier_Count", "Motif", "Inheritance", "Total_Loci")
colnames(XR_XY_result) <- c("Id", "Pathogenic_Count", "Motif", "Inheritance", "Total_Loci")
colnames(XD_result) <- c("Id", "Pathogenic_Count", "Motif", "Inheritance", "Total_Loci")

XR_XY_result$Id <- paste0("XY_", XR_XY_result$Id)
XR_XX_result$Id <- paste0("XX_", XR_XX_result$Id)

AD_result <- transform(AD_result, Carrier_Count = NA)
XR_XY_result <- transform(XR_XY_result, Carrier_Count = NA)
XD_result <- transform(XD_result, Carrier_Count = NA)

# combine data frames
combined_df <- do.call(rbind, list(AD_result, AR_result, XR_XX_result, XR_XY_result, XD_result))

combined_df$pathogenic_percent <- combined_df$Pathogenic_Count/combined_df$Total_Loci*100

combined_df$carrier_percent <- combined_df$Carrier_Count/combined_df$Total_Loci*100

combined_df <- calculate_ci(combined_df, successes_col = "Pathogenic_Count", trials_col = "Total_Loci", conf_level = 0.95)
combined_df <- calculate_ci(combined_df, successes_col = "Carrier_Count", trials_col = "Total_Loci", conf_level = 0.95)


combined_df <- combined_df %>%
  rename(gene = Id)

```

```{r comparison plot}

combined_df <- left_join(combined_df, STR_table %>% dplyr::select(gene, Prevalence), by = "gene") 


#where we don't know or have sex-specific prevalence
combined_df["Prevalence"][combined_df["Prevalence"] == ''] <- NA
combined_df$Prevalence[combined_df$gene == "XY_AFF2"] <- 2/50000
combined_df$Prevalence[combined_df$gene == "XY_AR"] <-  1/30000
combined_df$Prevalence[combined_df$gene == "XY_DMD"] <- 4.8/100000


combined_df$prevalence_dec <- sapply(combined_df$Prevalence, function(fraction) eval(parse(text = fraction)))

```


``` {r plots for IPG vs prev }

ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_point(data = filter(combined_df), alpha = 0.5) + # Plot all genes except HTT with lower alpha
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
    geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.5) +  # Add error bars for confidence intervals
    geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage")


ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.5) +  # Add error bars for confidence intervals
  geom_point(data = filter(combined_df, gene != "HTT"), alpha = 0.5) + # Plot all genes except HTT with lower alpha
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  coord_cartesian(ylim = c(1e-10, 5), xlim = c(1e-10, 5)) +
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  labs(title = "Comparison of Inferred Pathogenic Genotype from gNomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_y_log10() + scale_x_log10()


ggplot(combined_df, aes((x = prevalence_dec * 100) + 1e-10, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.8) +  
  geom_errorbar(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "gray", alpha = 0.3) +
  geom_point(data = filter(combined_df, gene != "HTT"), alpha = 0.8) + # Plot all genes except HTT with lower alpha
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  coord_cartesian(ylim = c(1e-10, 5), xlim = c(1e-5, 5)) +
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_y_log10() + scale_x_log10()

ggplot(subset(combined_df, !is.na(combined_df$prevalence)), 
       aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5)) + scale_y_break(c(0.65, 4), scales = "free")

```


``` {r options for side panel }
ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_point(data = filter(combined_df), alpha = 0.5) + # Plot all genes except HTT with lower alpha
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), width = 0.05, color = "black", alpha = 0.3) +  # Add error bars for confidence intervals
  geom_errorbar(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), width = 0.05, color = "black", alpha = 1) +  # Add error bars for confidence intervals
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage")

total_motif_match$gene <- factor(total_motif_match$gene, 
                                 levels = combined_df$gene[order(combined_df$pathogenic_percent)])

ggplot(data = filter(total_motif_match, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), aes(x = factor(gene), y = Allele2LowerBound, fill = Inheritance)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 0, bw = 0.03) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.4, xmax = as.numeric(factor(gene)) + 0.4, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.4, xmax = as.numeric(factor(gene)) + 0.4, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.4, xmax = as.numeric(factor(gene)) + 0.4, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1, bw = 0.03) +
  labs(title = "Allelic Distribution by Gene",
       x = "Gene",
       y = "Allele2 (repeat units)") +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45)) 


# Assuming total_motif_match is your dataframe and it's already filtered as needed
ggplot(data = filter(total_motif_match, gene %in% c("HTT", "XY_DMD", "ATXN8OS", "TCF4")), 
       aes(x = Allele2LowerBound+0.1, y = factor(gene), fill = Inheritance)) + 
  geom_density_ridges(scale = 0.6, alpha = 0) +
  geom_rect(aes(ymin = as.numeric(factor(gene)), ymax = as.numeric(factor(gene)) + 0.7, 
                xmin = normal_min, xmax = normal_max), fill = "#00c7d5") +
  geom_rect(aes(ymin = as.numeric(factor(gene)), ymax = as.numeric(factor(gene)) + 0.7, 
                xmin = intermediate_min, xmax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(ymin = as.numeric(factor(gene)), ymax = as.numeric(factor(gene)) + 0.7, 
                xmin = pathogenic_min, xmax = pathogenic_max), fill = "#FD6853") +
  geom_density_ridges(scale = 0.6) + scale_x_log10() + labs(title = "Allelic Distribution by Gene",
                                                            y = "Gene",
                                                            x = "Allele2 (repeat units)") +
  theme_minimal() + coord_cartesian(xlim = c(1, 1000))



```



```{r scraps }

ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_point(data = filter(combined_df, gene != "HTT"), alpha = 0.5) + # Plot all genes except HTT with lower alpha
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  geom_errorbar(aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.5) +  # Add error bars for confidence intervals
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_y_log10() + scale_x_log10()

ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.8) +  # Add error bars for confidence intervals
  geom_point(data = filter(combined_df, gene != "HTT"), alpha = 0.8) + # Plot all genes except HTT with lower alpha
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  coord_cartesian(ylim = c(1e-10, 5), xlim = c(1e-5, 5)) +
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_y_log10() + scale_x_log10()


ggplot(combined_df, aes(x = prevalence_dec * 100, y = pathogenic_percent + 1e-10, color = gene)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "black", alpha = 0.5) +  
  geom_errorbar(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.05, color = "gray", alpha = 0.3) +
  geom_point(data = filter(combined_df, gene != "HTT"), alpha = 0.5) + # Plot all genes except HTT with lower alpha
  geom_point(data = filter(combined_df, gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4")), alpha = 1) +    # Plot HTT with full alpha
  coord_cartesian(ylim = c(1e-10, 5), xlim = c(1e-10, 5)) +
  scale_color_manual(values = c("HTT" = "red", "TCF4" = "blue", "XY_DMD" = "orange", "ATXN8OS" = "purple"), 
                     na.value = "gray") + 
  theme_minimal() +
  labs(title = "Comparison of Inferred Pathogenic Genotype from gnomAD and Literature Known Prevalence",
       x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_y_log10() + scale_x_log10()


ggplot(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), aes(x = (prevalence_dec * 100)+ 1e-10, y = pathogenic_percent + 1e-10)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, !(gene %in% c("XY_DMD", "ATXN8OS", "TCF4"))), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.0005, color = "gray", alpha = 0.8) +
  geom_point(data = filter(combined_df, !(gene %in% c("XY_DMD", "ATXN8OS", "TCF4"))), alpha = 1) +    # Plot HTT with full alpha
  theme_minimal() +
    geom_text_repel(data = filter(combined_df, !(gene %in% c("XY_DMD", "ATXN8OS", "TCF4"))), aes(label = gene), 
                  box.padding = 0.5, segment.color = "grey50", segment.size = 0.2, size = 4,
                  nudge_y = 0, nudge_x = 0) +
  labs(x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage")



ggplot(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), aes(x = (prevalence_dec * 100)+ 1e-10, y = pathogenic_percent + 1e-10)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.0005, color = "black", alpha = 1) +
  geom_point(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), alpha = 1) +    # Plot HTT with full alpha
  theme_minimal() +
    geom_text_repel(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), aes(label = gene), 
                  box.padding = 0.5, segment.color = "grey50", segment.size = 0.2, size = 4,
                  nudge_y = 0, nudge_x = 0) +
  labs(x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage") + scale_x_log10() + scale_y_log10()

ggplot(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), aes(x = (prevalence_dec * 100)+ 1e-10, y = pathogenic_percent + 1e-10)) +
  geom_abline(intercept = 0, linetype = 'longdash', color = 'lightgrey') +
  geom_errorbar(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), 
                aes(ymin = 1e-10 + Pathogenic_Count_lower_ci, 
                    ymax = 1e-10 + Pathogenic_Count_upper_ci), 
                width = 0.0005, color = "black", alpha = 1) +
  geom_point(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), alpha = 1) +    # Plot HTT with full alpha
  theme_minimal() +
    geom_text_repel(data = filter(combined_df, !(gene %in% c("HTT","XY_DMD", "ATXN8OS", "TCF4"))), aes(label = gene), 
                  box.padding = 0.5, segment.color = "grey50", segment.size = 0.2, size = 4,
                  nudge_y = 0, nudge_x = 0) +
  labs(x = "Disease Prevalence",
       y = "Inferred Pathogenic Genotype Percentage")


```

``` {r plot, fig.width=15,fig.height=20}

ggplot(combined_df, aes(x = reorder(Id,percent), y = percent, color = Inheritance)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") + 
  ggtitle("Pathogenic % with Inheritance Logic, 95% CI") + facet_wrap(~Motif, ncol=2, scales = "free")


ggplot(combined_df, aes(x = reorder(Id,percent), y = percent, color = Motif)) +
  geom_point() +
  geom_errorbar(aes(ymin = lower_ci, ymax = upper_ci), width = 0.2) +
  xlab("Id") +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8)) +
  ylab("Pathogenic Percentage") + 
  ggtitle("Pathogenic % with Inheritance Logic, 95% CI") + facet_wrap(~Inheritance, ncol=2, scales = "free")

```


## Detour with Confidence Regions
"First, split it into two columns so we can plot it. I can think of a few plots that might be useful. Probably to start with, does the width of the confidence interval increase with allele2 size?"

```{r sanitycheck}
table(gnomADSTRcalls$Allele2LowerBound <= gnomADSTRcalls$Allele2 & gnomADSTRcalls$Allele2 <= gnomADSTRcalls$Allele2HigherBound)

table(gnomADSTRcalls$Allele2LowerBoundOTR <= gnomADSTRcalls$Allele2UsingOfftargetRegions & gnomADSTRcalls$Allele2UsingOfftargetRegions <= gnomADSTRcalls$Allele2HigherBoundOTR)

```

```{r confinterveal, fig.width=15,fig.height=100}

ggplot(gnomADSTRcalls, aes(x = Allele2, y = (Allele2HigherBound - Allele2LowerBound), col=Id)) +
  geom_point() + theme(legend.position = "none") + facet_wrap(~Id, ncol=2, scales = "free") +
  labs(y = "Width of Allele2 Confidence Interval", x = "Allele2")
```


```{r confinterveal spread out, fig.width=15,fig.height=100}
# ggplot(gnomADSTRcalls, aes(x = gnomADSTRcalls$Allele2, y = (gnomADSTRcalls$Allele2HigherBound - gnomADSTRcalls$Allele2LowerBound), col=Id)) +
#   geom_point() + theme(legend.position = "none") + facet_grid(~Id) + 
#   labs(y = "Width of Allele2 Confidence Interval", x = "Allele2")
```




# Do this after motifs have been separated. Then use one of the motifs sizes as the nchar bit. May need to finess RFC1 because it can have different motif sizes

```{r harrietplot}
# Combine Allele1 and Allele2

gnomADSTRcalls$RefAllele = round((gnomADSTRcalls$End - gnomADSTRcalls$Start_0based)/nchar(gnomADSTRcalls$Motif))

keep.cols = c('PcrProtocol', 'Age', 'Id', 'Motif', 'RefAllele')
allele1and2 = rbind(gnomADSTRcalls[,keep.cols], gnomADSTRcalls[,keep.cols])
allele1and2$Allele_withoutOTR = c(gnomADSTRcalls$Allele1, gnomADSTRcalls$Allele2)
allele1and2$Allele_withOTR = c(gnomADSTRcalls$Allele1UsingOfftargetRegions, gnomADSTRcalls$Allele2UsingOfftargetRegions)
allele1and2$Age[allele1and2$Age == 'age_not_available'] = NA

ggplot(subset(allele1and2, !is.na(Allele_withoutOTR)), 
              aes(x = Allele_withoutOTR,
               fill = Allele_withoutOTR == RefAllele)) + 
  geom_bar() + 
  facet_wrap(~Id, ncol=3, scales = "free") + 
  labs(x = 'Allele 1 or 2 withoutOTR')
ggsave('Allele_withoutOTR_isref.pdf', height = 50, width = 20, limitsize = FALSE)

ggplot(subset(allele1and2, !is.na(Allele_withOTR)), 
       aes(x = Allele_withOTR,
           fill = Allele_withOTR == RefAllele)) + 
  geom_bar() + 
  facet_wrap(~Id, ncol=3, scales = "free") + 
  labs(x = 'Allele 1 or 2 withOTR')
ggsave('Allele_withOTR_isref.pdf', height = 50, width = 20, limitsize = FALSE)


```

# Visualize things
"We should go take a look at some of those DMD and TCF4 calls to make sure they look real. Would you be able to output the image file names for a few of the largest ones?" from Harriet
Here's how to view them:
In your browser...
https://gnomad.broadinstitute.org/reads/gnomad_r3/short_tandem_repeats/[locus]/[filename].svg
For example:
https://gnomad.broadinstitute.org/reads/gnomad_r3/short_tandem_repeats/YEATS2/4b8c0a9ee270639517caa7472fe90ae9ff1.svg
Here's how to interpret the images:
https://gnomad.broadinstitute.org/news/2022-01-the-addition-of-short-tandem-repeat-calls-to-gnomad/#supplemental-details-f[â€¦]ing-read-visualizations 

``` {r visualize}
library(clipr)
# get largest allele2 from TCF4 to clipboard
TCF4_gnomAd <- subset(gnomADSTRcalls, Id == "TCF4")
TCF4_gnomAd <- TCF4_gnomAd[order(-TCF4_gnomAd$Allele2),]
clipr::write_clip(head(TCF4_gnomAd))

DMD_gnomAd <- subset(gnomADSTRcalls, Id == "DMD" & Sex == "XY")
DMD_gnomAd <- DMD_gnomAd[order(-DMD_gnomAd$Allele1),]
clipr::write_clip(head(DMD_gnomAd))

#https://gnomad.broadinstitute.org/reads/gnomad_r3/short_tandem_repeats/DMD/c7c0bdc287555a02382f282043126a5ce60.svg


$https://gnomad.broadinstitute.org/reads/gnomad_r3/short_tandem_repeats/DMD/[filename].svg

$https://gnomad.broadinstitute.org/reads/gnomad_r3/short_tandem_repeats/TCF4/[filename].svg


subset_ids_DMD <- DMD_gnomAd$PublicSampleId[DMD_gnomAd$AlleleLargest > 59]
print(subset_ids_DMD)


subset_ids_TCF4 <- TCF4_gnomAd$PublicSampleId[TCF4_gnomAd$AlleleLargest > 50]
print(subset_ids_TCF4)

intersect(subset_ids_DMD, subset_ids_TCF4)
union(subset_ids_DMD, subset_ids_TCF4)

```


# Fill by motif first, then separate by motif if needed. (edited) 

You can also embed plots, for example:

```{r pressure, echo=FALSE, fig.width=20}
valid_indices <- complete.cases(total$Allele2UsingOfftargetRegions, total$AlleleLargest)
fold_change <- total$Allele2UsingOfftargetRegions[valid_indices] / total$AlleleLargest[valid_indices]

# Create a new dataframe with fold_change and Id columns
df <- data.frame(fold_change = fold_change,
                 Id = total$Id[valid_indices],
                 PublicProjectId = total$PublicProjectId[valid_indices],
                 PublicSampleId = total$PublicSampleId[valid_indices],
                 ReferenceRegion = total$ReferenceRegion[valid_indices],
                 AlleleLargest = total$AlleleLargest[valid_indices],
                 Allele2UsingOfftargetRegions = total$Allele2UsingOfftargetRegions[valid_indices],
                 ReadvizFilename = total$ReadvizFilename[valid_indices]
                 )


df_motif <- data.frame(fold_change = fold_change,
                 Id = total$Id[valid_indices],
                 PublicProjectId = total$PublicProjectId[valid_indices],
                 PublicSampleId = total$PublicSampleId[valid_indices],
                 ReferenceRegion = total$ReferenceRegion[valid_indices],
               
                 Allele2 = total$Allele2[valid_indices],
                 Allele2UsingOfftargetRegions = total$Allele2UsingOfftargetRegions[valid_indices],
                 motif_norm= total$motif_norm[valid_indices],
                 ReadvizFilename = total$ReadvizFilename[valid_indices]
                 )

#   Allele1 = total$Allele1[valid_indices],

# # Create histogram plot with color by Id
# ggplot(data = df, aes(x = fold_change, fill = Id)) +
#   geom_histogram(binwidth = 1) +
#   scale_fill_manual(values = rainbow(length(unique(df$Id)))) +
#   scale_y_log10() +
#   labs(x = "Fold Change (Allele2UsingOfftargetRegions / Allele2)",
#        y = "Frequency",
#        title = "Distribution of Fold Change") + xlim(0,60) +
#   theme_minimal()
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.



### 1000 Genomes Phase 3 (?X Illumina sequencing of 3202 individuals)

- Crams: `/scratch/ucgd/lustre/UCGD_Datahub/IRBs/proj_UCGDCollab/A1290-220303-VAR-Masked_1KG_Phase3_WGS/UCGD/GRCh38/Data/PolishedCrams`
- Ref genome: `/scratch/ucgd/lustre/common/data/Reference/GRCh38/human_g1k_v38_decoy_phix.fasta`


JIGV to do programmatic 

``` {r wrong motif?}
# Create the histogram plot
# ggplot(total[total$Id == "STARD7", ], aes(x = AlleleLargest, fill = motif_norm)) +
#   geom_histogram(binwidth = 1) +
#   scale_fill_discrete() +
#   labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = STARD7)") +
#   theme_minimal()

# ggplot(data = subset(total, Id == "STARD7" & AlleleLargest > 10),
#        aes(x = AlleleLargest, fill = motif_norm)) +
#   geom_histogram(binwidth = 1) +
#   scale_fill_discrete() +
#   labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = STARD7)") +
#   theme_minimal()

# ggplot(data = subset(total, Id == "STARD7"),
#        aes(x = Allele2UsingOfftargetRegions, fill = motif_norm)) +
#   geom_histogram(binwidth = 1) +
#   scale_fill_discrete() +
#   labs(x = "Allele2UsingOfftargetRegions", y = "Frequency", title = "Histogram of Allele2UsingOfftargetRegions (Id = STARD7)") +
#   theme_minimal() + ylim(0,3) + xlim(10,900)

ggplot(df_motif[df_motif$Id == "STARD7", ], aes(x = fold_change, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "Fold Change", y = "Frequency", title = "Histogram of Fold Change (Id = STARD7)") +
  theme_minimal()

ggplot(data = subset(df_motif, Id == "STARD7" & fold_change > 10),
       aes(x = fold_change, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() + 
  labs(x = "Fold Change", y = "Frequency", title = "Histogram of Fold Change < 10(Id = STARD7)") +
  theme_minimal()


ggplot(data = subset(total, Id == "STARD7" & motif_norm != "AAAAT"), aes(x = AlleleLargest, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest w/o Most Common Motif (Id = RFC1)") +
  theme_minimal() 


# ggplot(data = subset(total, Id == "BEAN1"), aes(x = AlleleLargest, fill = motif_norm)) +
#   geom_histogram(binwidth = 1) +
#   scale_fill_discrete() +
#   labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = BEAN1)") +
#   theme_minimal() 

ggplot(data = subset(total, Id == "BEAN1" & motif_norm != "AAAAT"), aes(x = AlleleLargest, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = BEAN1)") +
  theme_minimal() 


ggplot(data = subset(df_motif, Id == "BEAN1" & fold_change > 5),
       aes(x = fold_change, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "Fold Change", y = "Frequency", title = "Histogram of Fold Change > 5 (Id = BEAN1)") +
  theme_minimal()

ggplot(total[total$Id == "DAB1", ], aes(x = AlleleLargest, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = DAB1)") +
  theme_minimal()

ggplot(data = subset(df_motif, Id == "DAB1" & fold_change > 5),
       aes(x = fold_change, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "Fold Change", y = "Frequency", title = "Histogram of Fold Change > 5 (Id = DAB1)") +
  theme_minimal()

ggplot(total[total$Id == "RFC1", ], aes(x = AlleleLargest, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest (Id = RFC1)") +
  theme_minimal()

ggplot(data = subset(total, Id == "RFC1" & motif_norm != "AAAAG"), aes(x = AlleleLargest, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "AlleleLargest", y = "Frequency", title = "Histogram of AlleleLargest w/o Most Common Motif (Id = RFC1)") +
  theme_minimal() 

ggplot(data = subset(df_motif, Id == "RFC1" & fold_change > 10),
       aes(x = fold_change, fill = motif_norm)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "Fold Change", y = "Frequency", title = "Histogram of Fold Change > 10 (Id = RFC1)") +
  theme_minimal()
```

```{r Allele1 ~= Allele1OTR}

ggplot(data = gnomADSTRcalls[gnomADSTRcalls$Allele1 != gnomADSTRcalls$Allele1UsingOfftargetRegions, ], 
       aes(x = Allele1, fill = Id)) +
  geom_histogram(binwidth = 1) +
  scale_fill_discrete() +
  labs(x = "Allele1", y = "Frequency", title = "Histogram of Allele1 (Unequal to Allele1UsingOfftargetRegions)") +
  theme_minimal()


ggplot(data = gnomADSTRcalls[gnomADSTRcalls$Allele1 != gnomADSTRcalls$Allele1UsingOfftargetRegions, ], 
       aes(x = Allele1, y = Allele1UsingOfftargetRegions, color = Id)) +
  geom_jitter() +
  scale_fill_discrete() +
  labs(x = "Allele1", y = "Allele1OTR", title = "Allele1 (Unequal to Allele1UsingOfftargetRegions)") +
  theme_minimal()


ggplot(data = gnomADSTRcalls[gnomADSTRcalls$Allele1 != gnomADSTRcalls$Allele1UsingOfftargetRegions, ], 
       aes(x = Allele1, y = Allele1UsingOfftargetRegions, color = Id)) +
  geom_jitter() +
  scale_fill_discrete() +
  labs(x = "Allele1", y = "Allele1OTR", title = "Allele1 (Unequal to Allele1UsingOfftargetRegions)") +
  theme_minimal() + xlim(0,300) + ylim(0,300)

data = gnomADSTRcalls[gnomADSTRcalls$Allele1 != gnomADSTRcalls$Allele1UsingOfftargetRegions, ]
```



Here are TRGT (STR) calls on a bunch of long read samples that may overlap with some of the samples Ben sent
/uufs/chpc.utah.edu/common/HIPAA/u6026198/quinlan-shared/data-shared/datasets/HPRC/trgt


```{r harriet strchive plots}

STR_table_adjusted <- STR_table %>%
  mutate(
    normal_min = coalesce(normal_min, normal_max),
    pathogenic_max = coalesce(pathogenic_max, pathogenic_min)
  )

STR_table_adjusted <- STR_table_adjusted %>%
  mutate(
    normal_min = ifelse(gene == "MARCHF6", 0, normal_min),
    normal_max = ifelse(gene == "MARCHF6", 0, normal_max)
  )

STR_table_adjusted$norm_min_bp = STR_table_adjusted$normal_min * STR_table_adjusted$repeatunitlen
STR_table_adjusted$norm_max_bp = STR_table_adjusted$normal_max * STR_table_adjusted$repeatunitlen
STR_table_adjusted$int_min_bp = STR_table_adjusted$intermediate_min * STR_table_adjusted$repeatunitlen
STR_table_adjusted$int_max_bp = STR_table_adjusted$intermediate_max * STR_table_adjusted$repeatunitlen
STR_table_adjusted$path_min_bp = STR_table_adjusted$pathogenic_min * STR_table_adjusted$repeatunitlen
STR_table_adjusted$path_max_bp = STR_table_adjusted$pathogenic_max * STR_table_adjusted$repeatunitlen

ggplot(STR_table_adjusted, aes(x = gene)) +
  geom_linerange(aes(ymin = norm_max_bp, ymax = path_max_bp, color = "gray"), 
                 linewidth = 1.5, linetype = "dotted", alpha = 0.3) +
  geom_point(data = subset(STR_table_adjusted, int_min_bp == int_max_bp), aes(y = int_min_bp), shape = 16, color= "#E7B800", size = 2) +
  geom_point(data = subset(STR_table_adjusted, path_min_bp == path_max_bp), aes(y = path_min_bp), shape = 16, color = "#FC4E07", size = 2) +
  geom_point(data = subset(STR_table_adjusted, norm_min_bp == norm_max_bp), aes(y = norm_min_bp), shape = 16, color= "#00AFBB", size = 2) +
  geom_linerange(aes(ymin = norm_min_bp+0.9, ymax = norm_max_bp, color = 'Normal'), linewidth = 1.5) +
  geom_linerange(aes(ymin = int_min_bp, ymax = int_max_bp, color = 'Intermediate*'), linewidth = 1.5) +
  geom_linerange(aes(ymin = path_min_bp, ymax = path_max_bp, color = 'Pathogenic'), linewidth = 1.5) +
  scale_y_continuous(name = 'Allele size in base pairs', trans = 'log10') +
  scale_x_discrete(name = 'Disease') +
  scale_colour_manual(values = c('Normal' = '#00AFBB', 'Intermediate*' = '#E7B800', 'Pathogenic' = '#FC4E07'), 
                      breaks = c('Normal', 'Intermediate*', 'Pathogenic'),
                      name = 'Allele size') + 
  theme(panel.grid.major.x = element_line(color = 'lightgrey', linetype = 'longdash')) +
  coord_flip()

```

``` {r age of onset, fig.height = 10}

# Loci with GeneReview Specific to Repeat Expansion Disease (>70% of variants)
# genes_to_bold <- c(
#   "AR", "ATN1", "ATXN1", "ATXN10", "ATXN2", "ATXN3", "ATXN7", "ATXN8OS",
#   "C9orf72", "CACNA1A", "DAB1", "DMPK", "FMR1", "FXN", "HTT", "JPH3",
#   "PABPN1", "PHOX2B", "RFC1", "TBP", "TCF4", "CNBP", "CSTB", "FGF14"
# )
# 
# bold_labels <- function(labels) {
#   lapply(labels, function(label) {
#     if (label %in% genes_to_bold) {
#       return(bquote(bold(.(label))))
#     } else {
#       return(as.character(label))
#     }
#   })
# }

evidence <- data.frame(
  gene = c("AFF2", "AFF3", "AR", "ARX_1", "ARX_2", "ATN1", "ATXN1", "ATXN10", "ATXN2", "ATXN3",
    "ATXN7", "ATXN8OS", "BEAN1", "C9orf72", "CACNA1A", "CBL", "COMP", "DAB1", "DIP2B",
    "DMD", "DMPK", "FMR1", "FOXL2", "FXN", "GIPC1", "GLS", "HOXA13_1", "HOXA13_2",
    "HOXA13_3", "HOXD13", "HTT", "JPH3", "LRP12", "MARCHF6", "NOP56", "NIPA1", "NOTCH2NLC",
    "NUTM2B-AS1", "PABPN1", "PHOX2B", "POLG", "PPP2R2B", "PRDM12", "RAPGEF2", "RFC1",
    "RILPL1", "RUNX2", "SAMD12", "SOX3", "STARD7", "TBP", "TBX1", "TCF4", "TNRC6A",
    "XYLT1", "YEATS2", "ZIC2", "ZIC3", "ZNF713", "CNBP", "CSTB", "EIF4A3", "PRNP",
    "VWA1", "ABCD3", "FGF14", "ZFHX3", "THAP11"
    ),
  ind_obs = c("100 > x > 50","< 10","> 100","50 > x > 10","< 10","100 > x > 50",
               "> 100","100 > x > 50","> 100","> 100","> 100","> 100","> 100",
               "> 100","> 100","< 10","< 10","50 > x > 10","< 10","< 10","> 100","> 100",
               "50 > x > 10","> 100","50 > x > 10","< 10","< 10","< 10","< 10","< 10","> 100",
              "100 > x > 50","100 > x > 50","> 100","100 > x > 50","> 100","> 100","< 10","> 100",
              "> 100","> 100","> 100","< 10","< 10","> 100","50 > x > 10","< 10","> 100","< 10",
              "50 > x > 10","100 > x > 50","< 10","> 100","< 10","50 > x > 10","< 10","50 > x > 10",
              "< 10","< 10","> 100","> 100","50 > x > 10","> 100","50 > x > 10","< 10","> 100",
              "50 > x > 10","< 10"))

STR_table_evidence <- merge(STR_table, evidence,
                   by = "gene", all.x = TRUE)
```

``` {r plotting age of onset, fig.height = 10}

ggplot(subset(STR_table_evidence, !is.na(age_onset_min) & Inheritance != ''), 
       aes(x = reorder(gene, -age_onset_min), color = Inheritance)) +
  scale_x_discrete(name = 'Gene') +
  geom_linerange(aes(ymin = age_onset_min, ymax = age_onset_max), size = 3, alpha = 0.4) +
  geom_linerange(aes(ymin = typ_age_onset_min, ymax = typ_age_onset_max), size = 3) +
  geom_point(data = subset(STR_table, age_onset_min == age_onset_max), aes(y = age_onset_min), size = 2.5, 
             alpha = 1, shape = 17) +
  scale_y_continuous(name = 'Age of onset (years)') +
  geom_segment(aes(x = 1, y = 18, xend = 67, yend = 18), linetype = 'longdash', color = 'lightgrey') +
  coord_flip() + scale_y_continuous(breaks=c(0,18, 25, 50, 75, 100))

ggplot(subset(STR_table_evidence, STR_table_evidence$ind_obs == "< 10" & !is.na(age_onset_min) & Inheritance != ''), 
       aes(x = reorder(gene, -age_onset_min), color = Inheritance)) +
  scale_x_discrete(name = 'Gene') +
  geom_linerange(aes(ymin = age_onset_min, ymax = age_onset_max), size = 3, alpha = 0.4) +
  geom_linerange(aes(ymin = typ_age_onset_min, ymax = typ_age_onset_max), size = 3) +
  geom_point(data = subset(STR_table_evidence, STR_table_evidence$ind_obs == "< 10" 
                            & age_onset_min == age_onset_max), aes(y = age_onset_min), size = 2.5, 
             alpha = 1, shape = 17) +
  scale_y_continuous(name = 'Age of onset (years)') +
  geom_segment(aes(x = 1, y = 18, xend = 24, yend = 18), linetype = 'longdash', color = 'lightgrey') +
  coord_flip() + scale_y_continuous(breaks=c(0,18, 25, 50, 75, 100)) + 
  labs(title = "<10 Independent Observations", y = "Age of Onset (years)")

ggplot(subset(STR_table_evidence, STR_table_evidence$ind_obs == "50 > x > 10"& !is.na(age_onset_min) & Inheritance != ''), 
       aes(x = reorder(gene, -age_onset_min), color = Inheritance)) +
  scale_x_discrete(name = 'Gene') +
  geom_linerange(aes(ymin = age_onset_min, ymax = age_onset_max), size = 3, alpha = 0.4) +
  geom_linerange(aes(ymin = typ_age_onset_min, ymax = typ_age_onset_max), size = 3) +
  geom_point(data = subset(STR_table_evidence, STR_table_evidence$ind_obs == "50 > x > 10" 
                            & age_onset_min == age_onset_max), aes(y = age_onset_min), size = 2.5, 
             alpha = 1, shape = 17) +
  scale_y_continuous(name = 'Age of onset (years)') +
  geom_segment(aes(x = 1, y = 18, xend = 12, yend = 18), linetype = 'longdash', color = 'lightgrey') +
  coord_flip() + scale_y_continuous(breaks=c(0,18, 25, 50, 75, 100)) + 
  labs(title = "10 < x < 50 Independent Observations", y = "Age of Onset (years)")

ggplot(subset(STR_table_evidence, STR_table_evidence$ind_obs == "100 > x > 50" & !is.na(age_onset_min) & Inheritance != ''), 
       aes(x = reorder(gene, -age_onset_min), color = Inheritance)) +
  scale_x_discrete(name = 'Gene') +
  geom_linerange(aes(ymin = age_onset_min, ymax = age_onset_max), size = 3, alpha = 0.4) +
  geom_linerange(aes(ymin = typ_age_onset_min, ymax = typ_age_onset_max), size = 3) +
  geom_point(data = subset(STR_table_evidence, STR_table_evidence$ind_obs == "100 > x > 50" 
                            & age_onset_min == age_onset_max), aes(y = age_onset_min), size = 2.5, 
             alpha = 1, shape = 17) +
  scale_y_continuous(name = 'Age of onset (years)') +
  geom_segment(aes(x = 1, y = 18, xend = 8, yend = 18), linetype = 'longdash', color = 'lightgrey') +
  coord_flip() + scale_y_continuous(breaks=c(0,18, 25, 50, 75, 100)) + 
  labs(title = " 50 < x < 100 Independent Observations", y = "Age of Onset (years)")


ggplot(subset(STR_table_evidence, STR_table_evidence$ind_obs == "> 100" & !is.na(age_onset_min) & Inheritance != ''), 
       aes(x = reorder(gene, -age_onset_min), color = Inheritance)) +
  scale_x_discrete(name = 'Gene') +
  geom_linerange(aes(ymin = age_onset_min, ymax = age_onset_max), size = 3, alpha = 0.4) +
  geom_linerange(aes(ymin = typ_age_onset_min, ymax = typ_age_onset_max), size = 3) +
  geom_point(data = subset(STR_table_evidence, STR_table_evidence$ind_obs == "> 100" 
                            & age_onset_min == age_onset_max), aes(y = age_onset_min), size = 2.5, 
             alpha = 1, shape = 17) +
  scale_y_continuous(name = 'Age of onset (years)') +
  geom_segment(aes(x = 1, y = 18, xend = 26, yend = 18), linetype = 'longdash', color = 'lightgrey') +
  coord_flip() + scale_y_continuous(breaks=c(0,18, 25, 50, 75, 100)) + 
  labs(title = ">100 Independent Observations", y = "Age of Onset (years)")

# STR_table %>%
#   filter(age_onset_min < 1) %>%
#   summarise(number_of_genes = n())

```


### Pie charts
```{r pie charts}
# Update the type column based on the Mechanism value (handling case sensitivity, whitespace, and NA values)
STR_table$type2 <- ifelse(tolower(trimws(STR_table$Mechanism)) == "polyglutamine", "PolyQ",
                  ifelse(tolower(trimws(STR_table$Mechanism)) == "polyalanine", "PolyA", STR_table$type))

STR_table <- STR_table %>%
  mutate(type2 = ifelse(gene == 'NIPA1', 'PolyA', type2))

STR_table <- STR_table %>%
  mutate(type2 = ifelse(gene == 'POLG', 'PolyQ', type2))

STR_table <- STR_table %>%
  mutate(type2 = ifelse(gene == 'PRDM12', 'PolyA', type2))

STR_table <- STR_table %>%
  mutate(type2 = ifelse(gene == 'THAP11', 'PolyQ', type2))

# Create a summary of the data by type
pie_data <- table(STR_table$type2)

# Plot the pie chart with labels
pie(pie_data, main = "Pie Chart of STR_table by Type", labels = paste(names(pie_data), " ", pie_data))


pie_data <- table(STR_table$Inheritance)
pie(pie_data, main = "STRchive by Inheritance", labels = paste(names(pie_data), " ", pie_data))

```

```{r coding breakdown}

STR_table_coding <- subset(STR_table, type2 == "Coding" | type2 == "PolyA" | type2 == "PolyQ")

STR_table_coding <- STR_table_coding %>%
  mutate(type2 = ifelse(gene == 'ZFHX3', 'PolyG', type2))

STR_table_coding <- STR_table_coding %>%
  mutate(type2 = ifelse(gene == 'COMP', 'PolyD', type2))

STR_table_coding <- STR_table_coding %>%
  mutate(type2 = ifelse(gene == 'JPH3', 'Other', type2))

STR_table_coding <- STR_table_coding %>%
  mutate(type2 = ifelse(gene == 'PRNP', 'Other', type2))

# STR_table_coding <- STR_table_coding %>%
#   mutate(type2 = ifelse(gene == 'JPH3', 'Coding:Poly-A or PolyL', type2))
# 
# STR_table_coding <- STR_table_coding %>%
#   mutate(type2 = ifelse(gene == 'PRNP', 'Coding:P, G, Q', type2))

STR_table_coding <- STR_table_coding %>%
  mutate(type2 = ifelse(gene == 'VWA1', 'Other', type2))


pie_data <- table(STR_table_coding$type2)

# Generate 5 shades of blue
green_palette <- colorRampPalette(c("lightgreen", "darkgreen"))(5)


# Plot the pie chart with labels
pie(pie_data, col = green_palette, main = "Coding Subtypes", labels = paste(names(pie_data),"(",pie_data,")"))

```



### Type and Motif Length
```{r type and motif length}
library(RColorBrewer)

# Generate 5 shades of blue
blue_palette <- colorRampPalette(c("lightblue", "darkblue"))(5)


# Count the number of characters in pathogenic_motif_reference_orientation
STR_table$length_characters <- str_length(STR_table$pathogenic_motif_reference_orientation)

STR_table$length_characters[STR_table$length_characters > 6] <- ">6"

STR_table <- STR_table %>%
  mutate(type = ifelse(gene == 'POLG', 'Coding', type))

# Plot
ggplot(STR_table, aes(x = type, fill = as.character(length_characters))) +
  geom_bar() +
  labs(title = "STR Loci Regions and Motif Length",
       x = "Type",
       y = "Count") +
  scale_fill_manual(values = blue_palette, name = "Motif Length") +
  theme_minimal()


# ggplot(subset(STR_table_coding, STR_table_coding$type == "Coding"), aes(x = type2, fill = as.character(length_characters))) +
#   geom_bar() +
#   labs(title = "Coding Type", x = "Subtypes",
#        y = "Count") +
#   theme(axis.text.x=element_text(angle = -90, hjust = 0)) +
#   theme_minimal()


```


```{r roadblock plot}
color_labels <- c('black', 'black', 'black', 'black', 'black', 
                  'black', 'black', 'black', 'blue','black', 'black', 'black', 'black', 'black', 'black', 
                  'purple', 'lightpink', 'orange')

ggplot(combined_df, aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5))

ggplot(data = filter(combined_df, !is.na(combined_df$prevalence)), aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5)) +
  theme(axis.text.x = element_text(colour = color_labels))
  

ggplot(data = filter(combined_df, !is.na(combined_df$prevalence)), 
       aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5)) + 
   ylim(0,0.64) +
  theme(axis.text.x = element_text(colour = color_labels))


ggplot(data = filter(combined_df, !is.na(combined_df$prevalence)), 
       aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5)) + 
  scale_y_log10()
  theme(axis.text.x = element_text(colour = color_labels))

iris1 <- ggplot(subset(combined_df, !is.na(combined_df$prevalence) & combined_df$pathogenic_percent < 1), 
       aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5))

iris2 <- ggplot(subset(combined_df, !is.na(combined_df$prevalence) & combined_df$pathogenic_percent > 1), 
                aes(x = reorder(gene, pathogenic_percent), y = pathogenic_percent)) +
  geom_point(aes(y = prevalence_dec*100), fill = "blue", size = 3, alpha = 0.9, shape = 23) +
  geom_point(aes(y = pathogenic_percent), color = "red", size = 3, alpha = 0.9, shape = 20) +
  geom_errorbar(aes(ymin = Pathogenic_Count_lower_ci, ymax = Pathogenic_Count_upper_ci), width = 0.2, color = "black") +
  labs(x = "Gene", y = "Percentage") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =8),
        plot.title = element_text(hjust = 0.5))

cowplot::plot_grid(iris1, iris2, labels = "AUTO")

```


```{r, allelic distribution, fig.height = 15}

#motif subset
total_motif_match <- total_prev %>%
  filter(total_prev$motif_norm %in% total_prev$repeatunit_path_normalized_list)

total_motif_match$Id <- ifelse(total_motif_match$Inheritance == "XR", paste0(total_motif_match$Sex, "_", total_motif_match$Id), total_motif_match$Id)

# Reorder levels of Id based on pathogenic_percent
total_motif_match$Id <- as.character(total_motif_match$Id)
combined_df$gene <- as.character(combined_df$gene)

total_motif_match <- total_motif_match %>%
  mutate(normal_min = ifelse(Id == 'PRNP', 3.8, normal_min))

total_motif_match <- total_motif_match %>%
  mutate(pathogenic_max = ifelse(Id == 'PRNP', 7, pathogenic_max))

total_motif_match <- total_motif_match %>%
  mutate(normal_min = ifelse(Id == 'VWA1', 1.9, normal_min))

total_motif_match <- total_motif_match %>%
  mutate(normal_max = ifelse(Id == 'VWA1', 2.1, normal_max))

total_motif_match <- total_motif_match %>%
  mutate(pathogenic_max = ifelse(Id == 'VWA1', 7, pathogenic_max))

total_motif_match <- total_motif_match %>%
  mutate(normal_max = ifelse(Id == 'COMP', 5.8, normal_max))

#this is just to see it
total_motif_match <- total_motif_match %>%
  mutate(pathogenic_max = ifelse(Id == 'NUTM2B-AS1', 705, pathogenic_max))

total_motif_match <- total_motif_match %>%
  rename(gene = Id)


total_motif_match <- merge(total_motif_match, combined_df[, c("gene", "pathogenic_percent", "carrier_percent")],
                   by = "gene", all.x = TRUE)

# Now reorder the levels based on pathogenic_percent
total_motif_match$gene <- factor(total_motif_match$gene, 
                               levels = combined_df$gene[order(combined_df$pathogenic_percent)])

# Create the plot with colored points
#ALLELE2LOWER BOUND!
### I think I like this one
AD_text <- total_motif_match %>%
  filter(Inheritance == 'AD') %>%  
  group_by(gene) %>%
  summarise(avg_pathogenic_percent = mean(pathogenic_percent, na.rm = TRUE)) %>%
    ungroup()


# 'Normal' = '#00AFBB', 'Intermediate*' = '#E7B800', 'Pathogenic' = '#FC4E07'),
ggplot(total_motif_match[total_motif_match$Inheritance == 'AD', ], aes(x = factor(gene), y = Allele2LowerBound)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
  geom_violin(fill = "black", scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  labs(title = "Allelic Distribution by Gene",
       x = "Gene",
       y = "Allele2 (repeat units)") +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45)) +   
  annotate("text", 
         x = AD_text$gene, 
         y = 200,
         label = paste0(round(AD_text$avg_pathogenic_percent, 3), "%"),
         vjust = -0.5, size = 2)

# Now reorder the levels based on pathogenic_percent
total_motif_match$gene <- factor(total_motif_match$gene, 
                               levels = combined_df$gene[order(combined_df$carrier_percent)])


AR_text <- total_motif_match %>%
  filter(Inheritance == 'AR') %>%  
  group_by(gene) %>%
  summarise(avg_pathogenic_percent = mean(pathogenic_percent, na.rm = TRUE),
            carrier_percent = mean(carrier_percent, na.rm = TRUE)) %>%
  ungroup()

## happy with it


# order by genotype, then carrier next
ggplot(total_motif_match[total_motif_match$Inheritance == 'AR', ], aes(x = factor(gene), y = Allele2LowerBound)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), 
            fill = "lightblue", alpha = 0.3) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
    geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_violin(fill = "gray", scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  labs(title = "Allelic Distribution by Gene",
       x = "Gene",
       y = "Allele2 (repeat units)") +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45)) + 
  annotate("text", 
         x = AR_text$gene, 
         y = 140,
         label = ifelse(is.nan(AR_text$carrier_percent), 
                        paste0(round(AR_text$avg_pathogenic_percent, 3), "%"),
                        paste0(round(AR_text$avg_pathogenic_percent, 3), "% | ", 
                               round(AR_text$carrier_percent, 3), "%")),
         vjust = -0.5, size = 3)


# Now reorder the levels based on pathogenic_percent

total_motif_match$gene <- factor(total_motif_match$gene, 
                               levels = combined_df$gene[order(combined_df$pathogenic_percent)])

XR_text <- total_motif_match %>%
  filter(Inheritance == 'XR') %>%  
  group_by(gene) %>%
  summarise(avg_pathogenic_percent = mean(pathogenic_percent, na.rm = TRUE),
            carrier_percent = mean(carrier_percent, na.rm = TRUE)) %>%
  ungroup()

### good as is
ggplot(total_motif_match[total_motif_match$Inheritance == 'XR', ], aes(x = factor(gene), y = Allele2LowerBound)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
    geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_violin(fill = "gray", scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$gene))),
              alpha = 1) +
  labs(title = "Allelic Distribution by Gene",
       x = "Gene",
       y = "Allele2 (repeat units)") +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45)) + 
  annotate("text", 
         x = XR_text$gene, 
         y = 140,
         label = ifelse(is.nan(XR_text$carrier_percent), 
                        paste0(round(XR_text$avg_pathogenic_percent, 3), "%"),
                        paste0(round(XR_text$avg_pathogenic_percent, 3), "% | ", 
                               round(XR_text$carrier_percent, 3), "%")),
         vjust = -0.5, size = 3)


#good as is
ggplot(total_motif_match[total_motif_match$Inheritance == 'XD', ], aes(x = factor(gene), y = Allele2LowerBound)) +
  geom_violin(scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$Id))),
              alpha = 1) +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = normal_min, ymax = normal_max), fill = "#00c7d5") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = intermediate_min, ymax = intermediate_max), fill = "#FFCB01") +
  geom_rect(aes(xmin = as.numeric(factor(gene)) - 0.5, xmax = as.numeric(factor(gene)) + 0.5, 
                ymin = pathogenic_min, ymax = pathogenic_max+0.5), fill = "#FD6853") +
  labs(title = "Allelic Distribution by Disease ID",
       x = "Disease ID",
       y = "Allele2 (repeat units)") +
  geom_violin(fill = "black", scale = "width", width = 0.9, 
              position = position_dodge(width = 0.9 / length(unique(total$Id))),
              alpha = 1) +
  scale_y_log10() +
  theme_minimal() +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45)) + 
  facet_wrap(~Sex)


```





```{r unique motif counts}

total_prev <- subset(total, total$motif_norm != "CNG")

#let's get the motif information
STR_table_motif <- STR_table[c("gene", "benign_motif_reference_orientation",
                               "unknown_motif_reference_orientation")]


#normalize motifs
STR_table_motif$repeatunit_ref_normalized <- sapply(STR_table$reference_motif_reference_orientation, function(x) normalise_str(as.character(x)))
STR_table_motif$repeatunit_path_normalized <- sapply(STR_table$pathogenic_motif_reference_orientation, function(x) normalise_str(as.character(x)))
STR_table_motif$benign_motif_reference_orientation <- sapply(STR_table$benign_motif_reference_orientation, function(x) normalise_str(as.character(x)))
STR_table_motif$unknown_motif_reference_orientation <- sapply(STR_table$unknown_motif_reference_orientation, function(x) normalise_str(as.character(x)))


# account for CNG ambiguity
STR_table_motif <- STR_table_motif %>%
  mutate(repeatunit_path_normalized = ifelse(repeatunit_path_normalized == 'CNG', 'CAG,CCG,CGG,CTG', repeatunit_path_normalized))
STR_table_motif <- STR_table_motif %>%
  mutate(repeatunit_ref_normalized = ifelse(repeatunit_ref_normalized == 'CNG', 'CAG,CCG,CGG,CTG', repeatunit_ref_normalized))

#get all motif counts
#sometimes motif_norm (larger allele) and motif_norm_small aren't the same
mismatched_rows <- total_prev %>%
  filter(motif_norm != motif_norm_small)

mismatched_summary <- total_prev %>%
  filter(motif_norm != motif_norm_small) %>%
  group_by(Id, motif_norm, motif_norm_small) %>%
  reframe(count = n())

#so, lets count both
total_prev_with_row_number <- total_prev %>%
  mutate(row_number = row_number())

# Convert motif_norm and motif_norm_small to separate rows, keeping only Id, row number, and the motifs
#you don't need row number but it's a good sanity check I think
motifs_by_allele <- total_prev_with_row_number %>%
  select(Id, motif_norm, motif_norm_small, row_number) %>%
  pivot_longer(cols = c(motif_norm, motif_norm_small), names_to = "motif_type", values_to = "motif")

# Count unique motifs by grouping by motif and Id
motif_counts <- motifs_by_allele %>%
  group_by(Id, motif) %>%
  reframe(MotifCount = n()) %>%
  pivot_wider(names_from = motif, values_from = MotifCount, values_fill = 0)

motif_counts_table <- motif_counts %>%
  mutate(MotifCount = as.character(MotifCount),  # Convert MotifCount to character
         MotifCount = paste(motif, MotifCount, sep = ": ")) %>%
  group_by(Id) %>%
  summarise(MotifCount = paste(MotifCount, collapse = ", ")) %>%
  ungroup()

#get unique motif counts
unique_motif_counts <- motifs_by_allele %>%
  group_by(Id) %>%
  reframe(UniqueMotifCount = n_distinct(motif),
          motif = paste0("", paste(unique(motif), collapse = ","), ""))

# more than 1 is more interesting
unique_motif_counts <- unique_motif_counts %>%
  filter(UniqueMotifCount != 1)

unique_motif_counts_long <- unique_motif_counts %>%
  mutate(motif = strsplit(as.character(motif), ",")) %>%
  unnest(motif)

unique_motif_counts_long <- unique_motif_counts_long %>%
  rename(gene = Id)

##merge it
unique_motif_counts_long <- merge(unique_motif_counts_long, STR_table_motif)


unique_motif_counts_long$classification <- apply(unique_motif_counts_long, 1, function(row) {
  m <- row['motif']
  if (any(m == unlist(strsplit(row['benign_motif_reference_orientation'], ',')))) {
    return("benign")
  } else if (any(m == unlist(strsplit(row['unknown_motif_reference_orientation'], ',')))) {
    return("unknown")
  } else if (any(m == unlist(strsplit(row['repeatunit_path_normalized'], ',')))) {
    return("path")
  } else if (any(m == unlist(strsplit(row['repeatunit_ref_normalized'], ',')))) {
    return("ref")
  } else {
    return("NEW_MOTIF?")
  }
})


# Reorder levels of classification factor
unique_motif_counts_long$classification <- factor(unique_motif_counts_long$classification,
                                                  levels = c("path", "unknown", "benign", "ref"))

# Plot with reordered levels
ggplot(unique_motif_counts_long, aes(x = gene, fill = classification)) +
  geom_bar() +
  labs(x = "Gene", y = "Unique Motif Count", fill = "Classification") +
  theme_minimal() +
  scale_fill_manual(values = c("benign" = "#00c7d5", "ref" = "darkblue", "unknown" = "lightgray", "path" = "#FD6853"),
                    labels = c("benign"= "Benign",
                               "ref" = "Reference",
                               "unknown" = "Unknown",
                               "path" = "Pathogenic"))




```


``` {r rfc1}
 RFC1_motif_counts <- subset(motifs_by_allele, motifs_by_allele$Id == 'RFC1')

 RFC1_motif_counts <- RFC1_motif_counts %>%
   rename(gene = Id)

 RFC1_motif_counts_motif_info <- merge(RFC1_motif_counts, STR_table_motif, by = "gene")
 
 
 RFC1_motif_counts_motif_info$classification <- apply(RFC1_motif_counts_motif_info, 1, function(row) {
  m <- row['motif']
  if (any(m == unlist(strsplit(row['repeatunit_ref_normalized'], ',')))) {
    return("ref")
  } else if (any(m == unlist(strsplit(row['unknown_motif_reference_orientation'], ',')))) {
    return("unknown")
  } else if (any(m == unlist(strsplit(row['repeatunit_path_normalized'], ',')))) {
    return("path")
  } else if (any(m == unlist(strsplit(row['benign_motif_reference_orientation'], ',')))) {
    return("benign")
  } else {
    return("NEW_MOTIF?")
  }
})



 # Reorder levels of classification factor
RFC1_motif_counts_motif_info$classification <- factor(RFC1_motif_counts_motif_info$classification,
                                                  levels = c("path", "unknown", "benign", "ref"))

 ggplot(RFC1_motif_counts_motif_info, aes(x=motif, fill = classification)) +
   geom_histogram(stat = "count") + scale_y_log10() +
   labs(title = "Motif Counts for RFC1", x = "Motif", y = "Frequency") +
   theme_minimal() +
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),  # Adjust size as needed
         axis.text.y = element_text(size = 14),  # Adjust size as needed
         axis.title.y = element_text(size = 16)) +
  scale_fill_manual(values = c("benign" = "#00c7d5", "ref" = "darkblue", "unknown" = "lightgray", "path" = "#FD6853"),
                    labels = c("benign"= "Benign",
                               "ref" = "Reference",
                               "unknown" = "Unknown",
                               "path" = "Pathogenic"))



 ggplot(RFC1_motif_counts_motif_info, aes(x=motif, fill = classification)) +
   geom_histogram(stat = "count") +
   labs(title = "Motif Counts for RFC1", x = "Motif", y = "Frequency") +
   theme_minimal() +
   theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),  # Adjust size as needed
         axis.text.y = element_text(size = 14),  # Adjust size as needed
         axis.title.y = element_text(size = 16)) +
  scale_fill_manual(values = c("benign" = "#00c7d5", "ref" = "darkblue", "unknown" = "lightgray", "path" = "#FD6853"),
                    labels = c("benign"= "Benign",
                               "ref" = "Reference",
                               "unknown" = "Unknown",
                               "path" = "Pathogenic"))





```

